---
title: "xv20230329_RPE1_pool_pathway assay"
author: "Xabier Vergara"
date: "2023-07-26"
output:
  html_document: default
  pdf_document: default
---
#Date: 28th August 2023
#Author: Xabier Vergara

#Aim: The aim of this file is to start with the first checks on the screening validation data. Answer questions like, do we need more sequencing depth, did all cell lines work, did all replicates work nicely?

#Conclusions: Overall, it looks better than last experiment. But, the conclusions remain very similar to the previous experiment. With the weird result on U2OS cells, that show positive correlation between balance and TIF.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(plotly)
library(ggbeeswarm)
library(pheatmap)
```

## Import data
```{r, warning=F, message=F}
# List files in mcclintock
pathway_assay_files <- list.files("/DATA/projects/DSBrepair/data/xv20230825_DSB_TRIP_revision/indelPCR_counts",pattern = '*[.]co', full.names = T) 

#Load sample info for the experiment
experiment_info <- readRDS("/DATA/projects/DSBrepair/data/xv20230825_DSB_TRIP_revision/xv20230825_sample_info.rds")

#Read and parse files
clone_pools_pathway_assay <- map_dfr(pathway_assay_files, function(x){
  read_delim(x, show_col_types = F) %>% 
    mutate(ID = str_extract(x,"7444_[:digit:].*(?=.count.table)"),
           outcome = paste(call, indel, sep = "_")) %>%
    left_join(experiment_info)
})
```


#Number of mapped IPRs with indel data (most optimistic)
```{r}
selected_pools <- c("RPE1_Low_1000","RPE1Deff_Low_1000","RPE1Proff_Low_250","U2OS_High_100")

#Import barcode sequences
#RPE cells
RPE_IPR <- read_delim("/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20230307_revision_figures/DSB_TRIP_mapping/xv20230519_RPE_pools_mapping.tsv") %>%
  select(name) %>%
  separate(name, into = c("barcode","iPCR_cell","iPCR_transfection","iPCR_complexity"))

U2OS_IPR <- read_delim("/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20230307_revision_figures/DSB_TRIP_mapping/xv20230519_U2OS_pools_mapping.tsv") %>%
  select(name) %>%
  separate(name, into = c("barcode","iPCR_cell","iPCR_transfection","iPCR_complexity"))

#Bind both
all_IPRs <- bind_rows(RPE_IPR, U2OS_IPR) %>% mutate(pool = paste(iPCR_cell, iPCR_transfection,iPCR_complexity, sep = "_")) %>% filter(pool %in% selected_pools)
```
#CONCLUSION: We are recovering the whole complexity for U2OS & PRO (input 400 ng). With DEF & RPE we are not capturing full complexity (input 200 ng).


```{r}
#Filter by replicate
indel_data <- clone_pools_pathway_assay %>% 
  dplyr::group_by(barcode,outcome, well, cell_line, gRNA, bio_rep) %>%
  dplyr::summarise(counts = sum(count, na.rm = F)) %>% left_join(all_IPRs %>% select(barcode, pool)) %>% na.omit()
```


#Explore LBR2 control and empty read counts
```{r}
#Filter control samples
control_samples <- indel_data %>%
  filter(gRNA %in% c("LBR2","halfLBR2","NTC"))

#Plot read numbers (with unedited reads)
ggplot(control_samples %>% filter(outcome %in% c("del_-7","ins_1","wt_0"))) + 
  geom_quasirandom(aes(gRNA, log10(counts), color = outcome)) + 
  facet_grid(bio_rep ~ cell_line)+
  theme_bw()

#Plot read numbers (only signature indels)
ggplot(control_samples %>% filter(outcome %in% c("del_-7","ins_1"))) + 
  geom_quasirandom(aes(gRNA, log10(counts), color = outcome)) + 
  facet_grid(bio_rep ~ cell_line) +
  theme_bw()

```
# Number of reads look OK, the same as editing efficiency

#Calculate total reads per sample and per barcode
```{r}
#Reads per sample
total_sample_reads <- indel_data %>%
  dplyr::group_by(well, cell_line, gRNA,bio_rep) %>%
  dplyr::summarise(total_sample_reads = sum(counts))

#Reads per barcode 
total_IPR_reads <- indel_data %>%
  dplyr::group_by(well,cell_line, gRNA, bio_rep, barcode) %>%
  dplyr::summarise(total_IPR_reads = sum(counts))

#Left_join
relative_barcode_reads <- left_join(total_sample_reads, total_IPR_reads) %>%
  mutate(IPR_read_freq = total_IPR_reads/total_sample_reads)

#Plot as a control
ggplot(relative_barcode_reads) + 
  geom_quasirandom(aes(gRNA, IPR_read_freq)) + 
  facet_grid(cell_line ~  bio_rep, scales = "free") + 
  theme_bw()

```

#Total sample read
```{r}
ggplot(relative_barcode_reads %>% 
         select(gRNA, total_sample_reads, bio_rep, cell_line) %>% 
         distinct()) +
  geom_quasirandom(aes(bio_rep,total_sample_reads, color= total_sample_reads < 100000)) +
  facet_wrap(~ cell_line) +
  theme_bw()
```

```{r}
ggplot(relative_barcode_reads %>% 
         select(gRNA, total_IPR_reads, bio_rep, cell_line) %>% 
         distinct()) +
  geom_bar(aes(gRNA, fill= total_IPR_reads < 250), position = "fill") +
  facet_wrap(~ cell_line, scales = "free") +
  theme_bw()

```


#Compute pathway balance in controls
```{r}
#Filter signature indels in control samples
control_balance_bio <- indel_data %>% filter( outcome %in% c("ins_1","del_-7","wt_0"))

#Look at read numbers
control_balance_bio_dcast <- control_balance_bio %>% reshape2::dcast(barcode + cell_line + gRNA + bio_rep ~ outcome, value.var = "counts", fill = 0, fun.aggregate = mean)

#Control_measurements (bio_rep)
pathway_metric_control_bio <- control_balance_bio_dcast %>%
  mutate(balance = `del_-7`/(ins_1 + `del_-7`),
         log2_bal = log2(`del_-7`/ins_1),
         TIF = 1 - (wt_0/(`del_-7`+ins_1+wt_0))) %>%
  na.omit()

#Filter by read number and plot noise level
read_count <- c(0,1,2,5,10,15,20,50,100,200,300)

filter_read_count_test <- map_dfr(read_count , function(x) {
  pathway_metric_control_bio %>%
    filter(`del_-7` >= x & ins_1 >= x) %>%
    mutate(read_filter = x)
})

#Number of IPRs in each
IPR_summary <- filter_read_count_test %>% 
  dplyr::group_by(cell_line, gRNA, read_filter, bio_rep) %>% 
  dplyr::summarize(IPR_n = n()) 

IPR_summary_dcast <- IPR_summary %>% 
  reshape2::dcast(cell_line + gRNA + bio_rep ~ read_filter, value.var  = "IPR_n")

#Plot loss of IPRs by each read_count filter

ggplot(IPR_summary) + 
  geom_col(aes(x = bio_rep,
               y = IPR_n,
               fill = fct_relevel(as.character(read_filter),as.character(read_count))),
           position = "dodge") +
  labs(fill = "read_filter") +
  facet_grid(cell_line~ gRNA, scales = "free") + 
  theme_bw()
```
#Conclusion: we always get some barcodes, but these are extremely low in RPE1 cells


#Check reproducibility among replicates
```{r}
#Calculate co-ocurrence per replicate
replicates_IPR <- filter_read_count_test %>%
  dplyr::group_by(barcode,cell_line, read_filter,gRNA) %>%
  dplyr::summarise(rep_n = n()) %>%
  ungroup() %>%
  dplyr::group_by(cell_line, read_filter,rep_n,gRNA) %>%
  dplyr::summarise(capture_IPR = n())

total_IPR_rep <- filter_read_count_test %>%
  select(cell_line, barcode, read_filter,gRNA) %>% 
  distinct() %>%
  dplyr::group_by(cell_line,read_filter,gRNA) %>%
  dplyr::summarise(total_IPR = n())

#Calculate frequency
freq_IPR_capture <- replicates_IPR %>% 
  left_join(total_IPR_rep) %>%
  mutate(freq_capture = capture_IPR/total_IPR)

#Absolute_numbers
IPR_reps <- replicates_IPR %>% reshape2::dcast(cell_line + rep_n ~ read_filter)

#IPRs in multiple reps
ggplot(replicates_IPR) + 
  geom_col(aes(x = fct_relevel(as.character(read_filter),as.character(read_count)),
               y = capture_IPR,
               fill = as.character(rep_n)),
           position = "dodge") +
  labs(fill = "IPR_reps") +
  facet_grid(cell_line ~ gRNA, scales = "free") + 
  theme_bw() + labs(x = "read_filter")

#Frequency
ggplot(freq_IPR_capture) + 
  geom_col(aes(x = fct_relevel(as.character(read_filter),as.character(read_count)),
               y = freq_capture,
               fill = as.character(rep_n)),
           position = "dodge") +
  labs(fill = "IPR_reps") +
  facet_grid(cell_line ~ gRNA , scales = "free") + 
  theme_bw() + labs(x = "read_filter")


```

#Calculate frequency in absolute number of IPRs
```{r}
#Total mapped IPR per pool
summary_mapped <- all_IPRs %>% 
  dplyr::group_by(iPCR_cell) %>% 
  dplyr::summarise(total_IPR = n()) %>%
  mutate(cell_line = case_when(iPCR_cell == "RPE1Deff" ~ "DEF",
                               iPCR_cell == "RPE1Proff" ~ "PRO",
                               T ~ iPCR_cell)) %>%
  select(cell_line, total_IPR)

#Calculate frequency out of all IPRs mapped
freq_IPR_capture_total <- replicates_IPR %>% 
  left_join(summary_mapped) %>%
  mutate(freq_capture = capture_IPR/total_IPR)

#Frequency
ggplot(freq_IPR_capture_total) + 
  geom_col(aes(x = fct_relevel(as.character(read_filter),as.character(read_count)),
               y = freq_capture,
               fill = as.character(rep_n)),
           position = "dodge") +
  labs(fill = "IPR_reps") +
  facet_wrap(~ cell_line) + 
  theme_bw() + labs(x = "read_filter")

#How many IPRs have at least n=2
n2_IPRs <- freq_IPR_capture_total %>% 
  filter(rep_n > 1) %>%
  dplyr::group_by(cell_line, read_filter,gRNA) %>%
  dplyr::summarise(IPR_n = sum(capture_IPR)) %>%
    left_join(summary_mapped) %>%
  mutate(freq_capture_n2 = IPR_n/total_IPR)
```

#Plot summary IPR number
```{r}
#Frequency
ggplot(n2_IPRs) + 
  geom_col(aes(x = gRNA,
               y = freq_capture_n2,
               fill = fct_relevel(as.character(read_filter),as.character(read_count))),
           position = "dodge") +
  labs(fill = "read_filter") +
  facet_wrap(~ cell_line) +
  theme_bw() +
  theme(legend.position = "top")
```

# CONCLUSION: IPRs captured per sample, look better than before! 


#Is pathway balance reproducible among replicates?
```{r, message=F, warning=F}
read_counts_filter_reprod <- filter_read_count_test %>%
  reshape2::dcast(barcode + cell_line + gRNA + read_filter ~ bio_rep, value.var = "balance")

#Calculate mean & sd
read_counts_filter_reprod_summary <- read_counts_filter_reprod %>% 
  rowwise() %>% 
  mutate(mean_bal = mean(c(R1,R2,R3),na.rm = T),
         sd_bal = sd(c(R1,R2,R3), na.rm = T),
         reps = 3 - sum(is.na(c(R1,R2,R3)))) %>%
  distinct()


#R1 vs. R2
ggplot(read_counts_filter_reprod_summary %>% filter(gRNA == "LBR2"), aes(R1,R2)) +
  geom_point(aes(color = as.character(reps))) + 
  geom_smooth(method = "lm") +
  ggpubr::stat_cor() +
  facet_grid(cell_line ~ read_filter, scales = "free") +
  theme_bw()

#R1 vs. R2
ggplot(read_counts_filter_reprod_summary, aes(R2,R3)) +
  geom_point(aes(color = as.character(reps))) + 
  geom_smooth(method = "lm") +
  ggpubr::stat_cor() +
  facet_grid(cell_line ~ read_filter, scales = "free") +
  theme_bw()

#R1 vs. R2
ggplot(read_counts_filter_reprod_summary, aes(R1,R3)) +
  geom_point(aes(color = as.character(reps))) + 
  geom_smooth(method = "lm") +
  ggpubr::stat_cor() +
  facet_grid(cell_line ~ read_filter, scales = "free") +
  theme_bw()

```

#Conclusion: To get high reproducibility in the results, we really need to be very stringent with the cut-off. I will try to repeat a similar filtering with total number of read cut-off and see how it looks.

#How does data look in this filter
```{R}
#Balance data
filtered_balance <- read_counts_filter_reprod_summary %>%
  filter(reps > 1) %>%
  select(barcode, cell_line, mean_bal, sd_bal, reps, read_filter, gRNA)

filtered_TIF <- read_counts_filter_reprod_summary_TIF %>%
  filter(reps > 1) %>%
  select(barcode, cell_line, mean_TIF = mean_bal, sd_TIF = sd_bal, reps, read_filter,gRNA)

filtered_measures <- left_join(filtered_balance, filtered_TIF)

```

#Plot balances and TIF
```{r}
#Plot balance
ggplot(filtered_measures) + 
  geom_quasirandom(aes(gRNA,mean_bal)) +
  facet_grid(cell_line ~ read_filter)
```

#Balance correlation between different samples
```{r}
#Prepare data frame
correlate_samples <- filtered_measures %>%
  reshape2::dcast(barcode + cell_line + read_filter ~ gRNA, value.var = "mean_bal")

#Plot LBR2 vs. halfLB2
ggplot(correlate_samples,aes(LBR2, halfLBR2)) +
  geom_point() +
  ggpubr::stat_cor() +
  geom_smooth(method = "lm") +
  facet_grid(cell_line ~ read_filter) +
  theme_bw()

#Plot LBR2 vs. halfLB2
ggplot(correlate_samples,aes(LBR2, NTC)) +
  geom_point() +
  ggpubr::stat_cor() +
  geom_smooth(method = "lm") +
  facet_grid(cell_line ~ read_filter) +
  theme_bw()
```
#Conclusion: Averaged data is reproducible! HalfLBR2 and LBR2 data are very similar and NTC (as expected by TIDE data) has lower MMEJ frequency mainly in RPE1 cells. However, we see a good correlation between LBR2 and NTC samples in RPE1 cells.

#Calculate delta log2 MMEJ:NHEJ (+1 > 10 reads & -7 > 10 reads)
```{r}
#Filter barcodes with reads
filtered_indel_data <- pathway_metric_control_bio %>%
  filter(`del_-7` > 50 & ins_1 > 50)

#extract control data
control_indel_data <- filtered_indel_data %>%
  filter(gRNA == "NTC") %>%
  select(barcode, cell_line, bio_rep, ntc_log2_bal = log2_bal)

#Prepare data frame
diff_data <- filtered_indel_data %>%
  left_join(control_indel_data) %>%
  na.omit() %>% 
  mutate(diff_balance = log2_bal - ntc_log2_bal)

#combine replicates
combined_diff_data <- diff_data %>%
  filter(bio_rep != "R1") %>%
  dplyr::group_by(barcode, cell_line, gRNA) %>%
  dplyr::summarise(mean_diff_balance = mean(diff_balance),
                   reps = n())


```

#Plot mean values per cell line
```{r}
mean_per_cell_line <- combined_diff_data %>%
  filter(reps > 1 & !gRNA %in% c("NTC","LBR2","halfLBR2","empty")) %>%
  dplyr::group_by(cell_line, gRNA) %>%
  dplyr::summarise(global_effect = mean(mean_diff_balance, na.rm = T),
                   sd_global = sd(mean_diff_balance, na.rm = T))

mean_per_cell_line_dcast <- mean_per_cell_line %>% 
  reshape2::dcast(gRNA ~ cell_line, value.var = "global_effect")

```

#Correlation by cell type
```{r} 
#KO_with protein quantity
highlight_prots <- c("36","38","44","54","60","68","72")

#RPE1 vs. PRO
ggplot(mean_per_cell_line_dcast, aes(RPE1, PRO)) + 
  geom_point(aes(color= gRNA %in% highlight_prots)) +
  geom_smooth(method = "lm") +
  ggpubr::stat_cor(method = "spearman") +
  geom_abline(linetype = 2) + 
  theme_bw()

#DEF vs. PRO
ggplot(mean_per_cell_line_dcast, aes(DEF, PRO)) + 
  geom_point(aes(color= gRNA %in% highlight_prots)) +
  geom_smooth(method = "lm") +
  ggpubr::stat_cor(method = "spearman") +
  geom_abline(linetype = 2) +
  theme_bw()

#U2OS vs. PRO
ggplot(mean_per_cell_line_dcast, aes(U2OS, PRO)) + 
  geom_point(aes(color= gRNA %in% highlight_prots)) +
  geom_smooth(method = "lm") +
  ggpubr::stat_cor(method = "spearman") +
  geom_abline(linetype = 2) +
  theme_bw()

```
#CONCLUSION: Global pathway balance effect are conserved across cell lines, in the validation set up.

#Calculate significance of globat values for volcano plot
```{r}
#Calculate wilcoxon test per gRNA
global_analysis_test_data <- combined_diff_data %>%
  filter(reps >1 & !gRNA %in% c("NTC","LBR2","halfLBR2","empty")) %>%
  dplyr::group_by(cell_line,gRNA) %>%
  wilcox_test(mean_diff_balance ~ 0) %>%
  ungroup() %>%
  dplyr::mutate(adj.p = p.adjust(p, method = "BH")) %>%
  left_join(mean_per_cell_line)

#Plot volcano plot
tmp <- ggplot(global_analysis_test_data, aes(global_effect, -log10(adj.p), gRNA = gRNA)) + 
  geom_point() +
  facet_wrap(~ cell_line) +
  theme_bw() +
  geom_vline(xintercept = 0, linetype = 2) +
  geom_hline(yintercept = -log10(0.001), linetype = 2)

ggplotly(tmp)

```
#CONCLUSION: I need higher quality data in RPE1. I will need this to say something about pathway balance. I need to retrieve higher number of IPRs.

#PRO & DEF: Global effect
```{r}
#Prepare data frame
PRO_DEF_data <- filtered_indel_data %>%
  filter(cell_line %in% c("PRO","DEF") & bio_rep != "R1") %>%
  dplyr::group_by(cell_line,gRNA, bio_rep) %>%
  dplyr::summarise(mean_balance = mean(log2_bal))

#Define control data PRO
control_PRO_indel_data <- PRO_DEF_data %>%
  filter(gRNA == "NTC" & cell_line == "PRO")  %>%
  ungroup() %>%
  select(bio_rep, PRO_log2_bal = mean_balance)

#combine replicates
PRO_combined_diff_data <- PRO_DEF_data %>%
  filter(!gRNA %in% c("empty","LBR2", "halfLBR2")) %>%
  left_join(control_PRO_indel_data) %>%
  mutate(diff_balance = mean_balance - PRO_log2_bal)

#Run statistics
PRO_global_analysis_test_data <- PRO_combined_diff_data %>%
  dplyr::group_by(cell_line,gRNA) %>%
  dplyr::summarise(global_effect = mean(diff_balance)) %>%
  reshape2::dcast(gRNA ~ cell_line) %>%
  mutate(int = DEF + 0.51,
         BRCA1 = -0.51,
         expected = PRO - 0.51)

#melt for plotting
melt_PRO_DEF_interaction <- PRO_global_analysis_test_data %>%
  reshape2::melt()

#Calculate melted interaction plot by replicate
PRO_global_analysis_test_by_rep <- PRO_combined_diff_data %>%
  reshape2::dcast(gRNA + bio_rep ~ cell_line, value.var = "diff_balance")

#Extract BRCA1 effect per replicate
BRCA1_effect <- PRO_global_analysis_test_by_rep %>%
  filter(gRNA == "NTC") %>%
  select(gRNA, bio_rep, BRCA1 = DEF)
 
```

#Plot BRCA1 and protein combination
```{r}
#Do they work on the same pathway or they have additive effects
ggplot(melt_PRO_DEF_interaction %>% 
         filter(variable %in% c("BRCA1","PRO","expected","DEF") & 
                  gRNA != "NTC")) + 
  geom_col(aes(fct_relevel(variable, order), value)) + 
  facet_wrap(~ gRNA) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1,
                                   vjust = 0.5))
```
#Export data
```{r}
write_rds(filtered_indel_data, file = "/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_screen/xv20230830_pathway_balance_50reads_per_indel.rds")

```
