---
title: "xv20220826_2_CCD_model"
author: "x.vergara"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# Script to generate figure #2 in the paper.

Fig. 2. CCDs of 89 DNA repair proteins. (A) Illustration of M- and N-synergy concepts. For chromatin feature – protein combinations with N-synergy (I in green) log2MMEJ:NHEJ scores increase as the chromatin feature levels increase. For combinations with M-synergy (II in brown) log2MMEJ:NHEJ scores decrease as the chromatin feature levels increase. For combinations with no synergy (III in black) log2MMEJ:NHEJ scores do not correlate with the individual chromatin features. (B) M-synergy example: linear fit (brown line) of RAD50 log2MMEJ:NHEJ scores with LMNB1 interaction. (C) No syn-ergy example: linear fit (black line) of MDC1 log2MMEJ:NHEJ scores with H2AFZ. (D) Heatmap of syn-ergy scores of all 89 proteins with significant CCDs. Proteins (columns) mentioned in the text are high-lighted in bold. Chromatin features (rows) are colored and ordered as in Fig. 1C. (E) Comparison of CCD effect sizes to global effect sizes of the proteins in (D). Global effect sizes are absolute values of global ∆log2MMEJ:NHEJ scores as calculated in Fig. 1B. CCD effect sizes are the predicted genome-wide dynamic range of ∆log2MMEJ:NHEJ values for each protein KO, as a function of variation in chromatin context  (see Methods). 

```{r functions, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(root.dir = '/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/rmarkdown/xv20210920_paper_scripts/short_version/knit_html/')

StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

#Change in.dir accordingly
in.dir <- "/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/"

#Check if output contain datetag
saveRDS_proof <- function(object, file) {
  document_path <- rstudioapi::getSourceEditorContext()$path %>% str_split(pattern = "/") %>% unlist()
  document_name <- document_path[length(document_path)]
  sys_time <- Sys.time()
  if (grepl(Date, file) == T) {
    saveRDS(object, file)
    #write(c(document_name, as.character(sys_time),"Output", correct_file), file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
  }
  else {
    print("Incorrect datetag!")
  }
}

#Read rds check through different versions
readRDS_proof <- function(path, pat) {
  full.path <- paste0(in.dir, path)
  document_path <- rstudioapi::getSourceEditorContext()$path %>% str_split(pattern = "/") %>% unlist()
  document_name <- document_path[length(document_path)]
  sys_time <- Sys.time()
  
  if(file.exists(full.path) == T){
    if(paste0(in.dir,path) != getwd()){
      setwd(paste0(in.dir, path))
  }
    file.list <- list.files(pattern = pat)
    correct_file <- str_sort(file.list, decreasing = T)[1]
    print(correct_file)
    readRDS(correct_file)
    #write(c(document_name,as.character(sys_time) ,"Input", correct_file), file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
    
  }
  else{
  setwd(path)
  file.list <- list.files(pattern = pat)
  correct_file <- str_sort(file.list, decreasing = T)[1]
  print(correct_file)
  readRDS(correct_file)
  #write(c(document_name, as.character(sys_time),"Input", correct_file),file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
  }
}
```

# Libraries
```{r libraries, warning=FALSE, message=FALSE}
# libraries:
library(tidyverse)
library(dendextend)
library(ggdendro)
library(grDevices)
library(pheatmap)
library(reshape2)
library(tibble)
library(ggpubr)
library(scales)
library(ggrepel)
library(ggdendro)
```

# Import data tables
```{r import , warning=FALSE, message=FALSE}
#setwd
setwd(in.dir)

#DNA repair CCDs
table_S3_CCDs  <- readRDS_proof("data/processed_data","Table_S7")

# heatmap chromatin order
heatmap.chromatin.order <- c("HDAC3","LMNB1","late_replicating","H3K9me2","H3K9me3","HDAC1","EZH2","H3K27me3","CTCF","SMC3","m5C","HDAC2","POL2","POL2AS2","H3K36me3","TTseq","Dam","H3K79me2","H2AFZ","DNAse","H3K4me3","H4K5acK8ac","H3K4me1","H3K27ac","H3K4me2")

#Import all different GO terms
GO_files <- list.files(path = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/import/GO_terms", pattern = "xv20231116.*txt", full.names = T)

#GO files per category
GO_files_all <- map_dfr(GO_files, function(x) {
  read.delim(x, header= F) %>% mutate(GO_ID = str_extract(x, "(?=human_GO).*(?<=.txt)"))
})

```


#GO:0000729 = DNA double-strand break processing
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_0000729_genes <- GO_files_all %>% filter(grepl("0000729", GO_ID)) %>% pull(V1) %>% unique()
GO_0000729 <- table_S3_CCDs %>% filter(gene %in% GO_0000729_genes)

# dcaast with selected
GO_0000729_dcast <- GO_0000729 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_0000729_heatmap <- pheatmap(t(GO_0000729_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 4)
GO_0000729_order <- rev(rownames(GO_0000729_dcast[GO_0000729_heatmap$tree_col[["order"]],]))

# This are the effects

# heatmap: All heatmaps together
ggplot(GO_0000729) + 
  geom_tile(aes(fct_relevel(gene,GO_0000729_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:0000729 DNA double-strand break processing")

```


#GO:1990918 = double-strand break repair involved in meiotic recombination
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_1990918_genes <- GO_files_all %>% filter(grepl("1990918", GO_ID)) %>% pull(V1) %>% unique()
GO_1990918 <- table_S3_CCDs %>% filter(gene %in% GO_1990918_genes)

# dcaast with selected
GO_1990918_dcast <- GO_1990918 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_1990918_heatmap <- pheatmap(t(GO_1990918_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 3)
GO_1990918_order <- rev(rownames(GO_1990918_dcast[GO_1990918_heatmap$tree_col[["order"]],]))

# This are the effects

# heatmap: All heatmaps together
ggplot(GO_1990918) + 
  geom_tile(aes(fct_relevel(gene,GO_1990918_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:1990918 DNA double-strand break processing")

```

GO:0000724 = double-strand break repair involved in homologus recombination
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_0000724_genes <- GO_files_all %>% filter(grepl("0000724", GO_ID)) %>% pull(V1) %>% unique()
GO_0000724 <- table_S3_CCDs %>% filter(gene %in% GO_0000724_genes)

# dcaast with selected
GO_0000724_dcast <- GO_0000724 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_0000724_heatmap <- pheatmap(t(GO_0000724_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 4)
GO_0000724_order <- rev(rownames(GO_0000724_dcast[GO_0000724_heatmap$tree_col[["order"]],]))

# This are the effects

# heatmap: All heatmaps together
ggplot(GO_0000724) + 
  geom_tile(aes(fct_relevel(gene,GO_0000724_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:0000724 = double-strand break repair involved in homologus recombination")

```

GO:0045002 = double-strnad break repair by single-strand annealing
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_0045002_genes <- GO_files_all %>% filter(grepl("0045002", GO_ID)) %>% pull(V1) %>% unique()
GO_0045002 <- table_S3_CCDs %>% filter(gene %in% GO_0045002_genes)

# dcaast with selected
GO_0045002_dcast <- GO_0045002 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_0045002_heatmap <- pheatmap(t(GO_0045002_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 3)
GO_0045002_order <- rev(rownames(GO_0045002_dcast[GO_0045002_heatmap$tree_col[["order"]],]))

# This are the effects

# heatmap: All heatmaps together
ggplot(GO_0045002) + 
  geom_tile(aes(fct_relevel(gene,GO_0045002_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:0045002 = double-strnad break repair by single-strand annealing")

```


GO:0006303 = double-strand break repair involved in non-homologous end joining
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_0006303_genes <- GO_files_all %>% filter(grepl("0006303", GO_ID)) %>% pull(V1) %>% unique()
GO_0006303 <- table_S3_CCDs %>% filter(gene %in% GO_0006303_genes & !is.na(chrom_feature))

# dcaast with selected
GO_0006303_dcast <- GO_0006303 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_0006303_heatmap <- pheatmap(t(GO_0006303_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 3)
GO_0006303_order <- rev(rownames(GO_0006303_dcast[GO_0006303_heatmap$tree_col[["order"]],]))

# This are the effects

# heatmap: All heatmaps together
ggplot(GO_0006303) + 
  geom_tile(aes(fct_relevel(gene,GO_0006303_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:0006303 = double-strand break repair involved in non-homologous end joining")

```

GO:0097681 = double-strand break repair via alternative nonhomologous end joining
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_0097681_genes <- GO_files_all %>% filter(grepl("0097681", GO_ID)) %>% pull(V1) %>% unique()
GO_0097681 <- table_S3_CCDs %>% filter(gene %in% GO_0097681_genes & !is.na(chrom_feature))

# dcaast with selected
GO_0097681_dcast <- GO_0097681 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_0097681_heatmap <- pheatmap(t(GO_0097681_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 3)
GO_0097681_order <- rev(rownames(GO_0097681_dcast[GO_0097681_heatmap$tree_col[["order"]],]))

# This are the effects

# heatmap: All heatmaps together
ggplot(GO_0097681) + 
  geom_tile(aes(fct_relevel(gene,GO_0097681_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:0097681 = double-strand break repair via alternative nonhomologous end joining")

```

GO:0097680 = double-strand break repair via classical nonhomologous end joining
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_0097680_genes <- GO_files_all %>% filter(grepl("0097680", GO_ID)) %>% pull(V1) %>% unique()
GO_0097680 <- table_S3_CCDs %>% filter(gene %in% GO_0097680_genes & !is.na(chrom_feature))

# dcaast with selected
GO_0097680_dcast <- GO_0097680 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_0097680_heatmap <- pheatmap(t(GO_0097680_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 3)
GO_0097680_order <- rev(rownames(GO_0097680_dcast[GO_0097680_heatmap$tree_col[["order"]],]))

# This are the effects

# heatmap: All heatmaps together
ggplot(GO_0097680) + 
  geom_tile(aes(fct_relevel(gene,GO_0097680_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:0097680 = double-strand break repair via alternative nonhomologous end joining")

```

GO:2001033 = negative regulation of double-strand break repair via nonhomologous end joining
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_2001033_genes <- GO_files_all %>% filter(grepl("2001033", GO_ID)) %>% pull(V1) %>% unique()
GO_2001033 <- table_S3_CCDs %>% filter(gene %in% GO_2001033_genes & !is.na(chrom_feature))

# dcaast with selected
GO_2001033_dcast <- GO_2001033 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")


# heatmap: All heatmaps together
ggplot(GO_2001033) + 
  geom_tile(aes(fct_relevel(gene,GO_0097680_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:2001033 = double-strand break repair via alternative nonhomologous end joining")

```


GO:2001034 = positive regulation of double-strand break repair via nonhomologous end joining
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_2001034_genes <- GO_files_all %>% filter(grepl("2001034", GO_ID)) %>% pull(V1) %>% unique()
GO_2001034 <- table_S3_CCDs %>% filter(gene %in% GO_2001034_genes & !is.na(chrom_feature))

# dcaast with selected
GO_2001034_dcast <- GO_2001034 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_2001034_heatmap <- pheatmap(t(GO_2001034_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 3)
GO_2001034_order <- rev(rownames(GO_2001034_dcast[GO_2001034_heatmap$tree_col[["order"]],]))




# This are the effects

# heatmap: All heatmaps together
ggplot(GO_2001034) + 
  geom_tile(aes(fct_relevel(gene,GO_2001034_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:2001034 = positive regulation of double-strand break repair via nonhomologous end joining")

```


GO:2001032 = regulation of double-strand break repair via nonhomologous end joining
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_2001032_genes <- GO_files_all %>% filter(grepl("2001032", GO_ID)) %>% pull(V1) %>% unique()
GO_2001032 <- table_S3_CCDs %>% filter(gene %in% GO_2001032_genes & !is.na(chrom_feature))

# dcaast with selected
GO_2001032_dcast <- GO_2001032 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_2001032_heatmap <- pheatmap(t(GO_2001032_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 3)
GO_2001032_order <- rev(rownames(GO_2001032_dcast[GO_2001032_heatmap$tree_col[["order"]],]))




# This are the effects

# heatmap: All heatmaps together
ggplot(GO_2001032) + 
  geom_tile(aes(fct_relevel(gene,GO_2001032_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:2001032 = regulation of double-strand break repair via nonhomologous end joining")

```




GO:2000780 = negative regulation of double-strand break repair
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_2000780_genes <- GO_files_all %>% filter(grepl("2000780", GO_ID)) %>% pull(V1) %>% unique()
GO_2000780 <- table_S3_CCDs %>% filter(gene %in% GO_2000780_genes & !is.na(chrom_feature))

# dcaast with selected
GO_2000780_dcast <- GO_2000780 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_2000780_heatmap <- pheatmap(t(GO_2000780_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 4)
GO_2000780_order <- rev(rownames(GO_2000780_dcast[GO_2000780_heatmap$tree_col[["order"]],]))


# This are the effects

# heatmap: All heatmaps together
ggplot(GO_2000780) + 
  geom_tile(aes(fct_relevel(gene,GO_2000780_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:2000780 = negative regulation of double-strand break repair")

```


GO:2000781 = positive regulation of double-strand break repair
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_2000781_genes <- GO_files_all %>% filter(grepl("2000781", GO_ID)) %>% pull(V1) %>% unique()
GO_2000781 <- table_S3_CCDs %>% filter(gene %in% GO_2000781_genes & !is.na(chrom_feature))

# dcaast with selected
GO_2000781_dcast <- GO_2000781 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_2000781_heatmap <- pheatmap(t(GO_2000781_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 4)
GO_2000781_order <- rev(rownames(GO_2000781_dcast[GO_2000781_heatmap$tree_col[["order"]],]))


# This are the effects

# heatmap: All heatmaps together
ggplot(GO_2000781) + 
  geom_tile(aes(fct_relevel(gene,GO_2000781_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:2000781 = positive regulation of double-strand break repair")

```

GO:2000779 = regulation of double-strand break repair
```{r Fig_2_D, warning=FALSE, message=FALSE}
# Separate data based on the significant IPRs that they have (based on z-score)
GO_2000779_genes <- GO_files_all %>% filter(grepl("2000779", GO_ID)) %>% pull(V1) %>% unique()
GO_2000779 <- table_S3_CCDs %>% filter(gene %in% GO_2000779_genes & !is.na(chrom_feature))

# dcaast with selected
GO_2000779_dcast <- GO_2000779 %>% reshape2::dcast(gene ~ chrom_feature, value.var = "CCD_synergy_score") %>% column_to_rownames(var = "gene")

#Heatmap
GO_2000779_heatmap <- pheatmap(t(GO_2000779_dcast), silent = T, clustering_method = "ward.D", cluster_rows = T, cutree_cols = 4)
GO_2000779_order <- rev(rownames(GO_2000779_dcast[GO_2000779_heatmap$tree_col[["order"]],]))


# This are the effects

# heatmap: All heatmaps together
ggplot(GO_2000779) + 
  geom_tile(aes(fct_relevel(gene,GO_2000779_order),
                fct_relevel(chrom_feature, heatmap.chromatin.order),
                fill = CCD_synergy_score)) +
  scale_fill_gradient2( low = "#8c510a",mid = "#f5f5f5", high = "#01665e",limits = c(-0.15,0.15), oob = squish, breaks = c(-0.1,0,0.1)) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "right") + 
  coord_fixed(expand = F,ratio = 1) + 
  xlab("DNA repair proteins with significant chromatin dependencies") + 
  ylab("Chromatin features") +
  ggtitle("GO:2000779 = regulation of double-strand break repair")

```

