---
title: "xv20230227_follow_up_library"
author: "Xabier Vergara"
date: "2023-02-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Follow up library design
in this markdown, I will design a library to validate the main findings on the screen.
Potential candidate list for genes to target
NHEJ_factors <- POLL, BRCA2, CHAF1A ('D'), BABAM2, UIMC1, BRCC3.
MMEJ_factors_TH <- RBBP8 ('D'), ATM, FANCD2, TOPBP1 ('E'), MDC1, RAD50, ATR ('E')
MMEJ_factors_K27 <- CHEK2, FANCM ('D'), FANCG, FAAP24, SMC5 ('E')
MMEJ_factors_euch <- RMI2 ('D'), BLM, PARP1
Dual <- BOD1L1


```{r libraries}
library(tidyverse)
library(ggbeeswarm)
```
#Functions
```{r}
#Function find reference kgID
kgID_reference_finding <- function(x) {
  kgID_summary <- x %>% 
  select(kgID, Exon.number) %>%
  distinct() %>%
  dplyr::group_by(kgID) %>% 
  dplyr::summarise(exon_counts = n())
  
  #max exons
  kgID_reference <- top_n(kgID_summary, 1, exon_counts) %>% pull(kgID)
  
  #Select the kgID to use: If overlap is big take one randomly
  kgID_to_use <- if (length(kgID_reference) > 1) {
  
  #Unique gRNAs
  unique_gRNAs <- x %>%
    filter(kgID %in% kgID_reference) %>%
    pull(gRNA) %>% unique()
  
  #All gRNAs
  all_gRNAs <- x %>%
    filter(kgID %in% kgID_reference) %>%
    pull(gRNA)
  
  #What's the overlap frequency
  overlap <- length(all_gRNAs)/(length(unique_gRNAs)*length(kgID_reference))
  print(overlap)

  if (overlap > 0.9) { #If overlap is high take only one
  kgID_reference[1]
  } else { #If overlap is not high keep both
    kgID_reference
  }
} else { #In case there is only one reference pick that one
  kgID_reference
}
}

#Summary gene table calculation (x = high_quality)
make_summary_table <- function(x,y) {
  #Number of gRNAs per exon
summary_gRNA_hs <- x %>% 
  dplyr::group_by(kgID, Exon.number) %>%
  dplyr::summarise(gRNA_n = n(),
                   common_gRNA_n = sum(gRNA %in% y),
                   Exon.length = Exon.end - Exon.start) %>%
  ungroup() %>%
  distinct()

#Calculate distance to previous and next exon
distance_exon <- map_dfr(unique(x$kgID), function(kg){
  one_kg <- x %>% filter(kgID == kg)
  map_dfr(unique(one_kg$Exon.number), function(ex){
    pos_exon <- which(unique(one_kg$Exon.number) == ex)
    prev_index <- unique(one_kg$Exon.number)[pos_exon - 1]
    next_index <- unique(one_kg$Exon.number)[pos_exon + 1]
    exon_start <- x %>% filter(Exon.number == ex) %>% pull(Exon.start) %>% unique()
    exon_end <- x %>% filter(Exon.number == ex) %>% pull(Exon.end) %>% unique()
    
    #Function for previous index
    if (is_empty(prev_index)) {
      prev_exon_end <- exon_start
    } else {
      prev_exon_end <- x %>% filter(Exon.number == prev_index) %>% pull(Exon.end) %>% unique()
    }
    
    #Function for next exon
    if (is_empty(next_index)) {
      next_exon_start <- exon_end
    } else {
      next_exon_start <- x %>% filter(Exon.number == next_index) %>% pull(Exon.start) %>% unique()
    }
    
    
    tibble(kgID = kg,
         Exon.number = ex,
         Dist.prev.exon = abs(exon_start - prev_exon_end),
         Dist.next.exon = abs(next_exon_start - exon_end))
  })
})

#Left_join
left_join(summary_gRNA_hs, distance_exon, by = c("kgID","Exon.number")) %>% distinct()
}

#Function to add one more exons
  add_exon_gRNA <- function(x,y){
    targetted_area <- x %>%
      filter(Exon.number == y)
    targetted_prev <- 0
    i <- 1
    while (targetted_prev < 5000){
      #Set_counter
      exon <- y - i
      #Add previous 
      previous_exon <- x %>% 
        filter(Exon.number == exon)
      #Bind columns
      targetted_area <- targetted_area %>% bind_rows(previous_exon)
      #Check if the condition is met
      targetted_prev <- sum(targetted_area$Exon.length) + sum(targetted_area$Dist.prev.exon)
      if (targetted_area$Dist.prev.exon[1] != 0) {
        #Add one to the counter
        i <- i + 1
      } else {
      break
      }
    }
    targetted_area
  }
```

#Import data tables: mouse and human
```{r pressure, warning=F, message=F}
#Write a long function that I could use to run iteratively
analysis_gRNA_indelPhi <- function(x) {
  #Load data
  print(paste("This is the summary of ",x,"gene:"))
  human_gRNA <- read.delim(file = paste0("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_library/indelphi_human/",x,"_hg38.csv"), sep = ",")
  mouse_gRNA <- read.delim(file = paste0("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_library/indelphi_mouse/",x,"_mm10.csv"), sep = ",")
  
  gRNAs_both_species <- intersect(human_gRNA$gRNA,mouse_gRNA$gRNA)

  
  #High quality gRNAs (Frameshift > 80%)
  high_quality_gRNAs <- human_gRNA %>%
  dplyr::group_by(kgID) %>%
  dplyr::mutate(rel_ex_number = Exon.number/max(Exon.number)) %>%
  ungroup() %>%
  filter(Frameshift.... > 80 & rel_ex_number < 0.2) %>%
  distinct()
  
  #Select_reference kgID
  reference_kgID <- kgID_reference_finding(high_quality_gRNAs)
  
  print(paste("For this analysis, we will use the gRNAs targetting the gene build",reference_kgID,"."))
  
  high_quality_gRNAs <- high_quality_gRNAs %>% filter(kgID %in% reference_kgID)
  
  #Calculate summary
  summary_table <- make_summary_table(high_quality_gRNAs, gRNAs_both_species)
  
  #Export summary table
write.table(summary_table,
            file = paste0("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_library/Indelphi_analysis/summary_table_",x,".txt"))
  
  #Exon with highest gRNA numbers
  highest_gRNA <- summary_table %>% 
  ungroup() %>% 
  top_n(1,gRNA_n)
  
  print(highest_gRNA)

#Exon with highest gRNA to target mouse and human
  highest_common <- summary_table %>% 
  ungroup() %>% 
  filter(common_gRNA_n > 1) %>%
  top_n(1,common_gRNA_n)

#Make a quick if to select starting exon number
 exons_to_target <- if (nrow(highest_common) > 0){
  add_exon_gRNA(summary_table, highest_common$Exon.number)
  } else {
  add_exon_gRNA(summary_table, highest_gRNA$Exon.number)
  }
 print(paste("I will design gRNA targetting the", paste(unique(exons_to_target$Exon.number),collapse = ","),"exons."))
 print(exons_to_target$Exon.number)

# Filter high quality gRNAs in targetted regions (save as txt file)
gRNA_to_select <- high_quality_gRNAs %>% 
  filter(Exon.number %in% exons_to_target$Exon.number) %>%
  mutate(common_gRNA = gRNA %in% gRNAs_both_species)

#Export all steps
write.table(gRNA_to_select,
            file = paste0("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_library/Indelphi_analysis/gRNAs_targetting_",x,"_complete.txt"),
            row.names = F, quote = F)

#Export FASTA for off targets
write.table(unique(gRNA_to_select$gRNA),
            file = paste0("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_library/Indelphi_analysis/gRNAs_targetting_",x,"_off_targets.txt"),
            row.names = F, quote = F, col.names = F)

#Printed summary
print(paste("There are a total of", sum(exons_to_target$gRNA_n), "gRNAs in the selected region and",sum(exons_to_target$common_gRNA_n), "target both species."))
print("===============================================")

}

#Perform a loop
files <- list.files("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_library/indelphi_human/")
gene_names <- gsub("_hg38.csv","",files)

for (i in gene_names) {
  analysis_gRNA_indelPhi(i)
}


```

#Plot different kgIDs
```{r}
# Plot kgID
ggplot(human_gRNA) + geom_point(aes(Cutsite.coordinate, kgID, color = kgID %in% kgID_to_use)) +
  theme_bw()
```
#Filter high quality gRNAs
```{r}

```

#Calculate summary table of gRNAs
```{r}

```

#Summary plots: Exons
```{r}
#number of gRNAs per exon per position
ggplot(summary_gene_hs) +
  geom_point(aes(as.numeric(Exon.number), gRNA_n,size = common_gRNA_n, color = common_gRNA_n > 0)) + 
  facet_wrap(~kgID) + theme_bw() + 
  theme(legend.position = "top") +
  coord_cartesian(ylim = c(0, max(summary_gene_hs$gRNA_n))) + xlab("Exon number") + ylab("Number of high quality gRNA per exon")

```



#How many gRNAs could be used in mouse cells?
```{r}
#Print
print(paste("There are a total of", sum(exons_to_target$gRNA_n), "gRNAs in the selected region and",sum(exons_to_target$common_gRNA_n), "target both species."))
```

#Select gRNAs and export them
```{r}


```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
