---
title: "xv20220712_S7_editing_efficiency"
author: "x.vergara"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---
# Script to generate Supplementary Figure 7

Fig. S7: Effect size and knock-out penetrance estimation.
```{r functions, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(root.dir = '/DATA/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/rmarkdown/xv20210920_paper_scripts/short_version/knit_html/')

StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

#Change in.dir accordingly
in.dir <- "/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/"

#Check if output contain datetag
saveRDS_proof <- function(object, file) {
  document_path <- rstudioapi::getSourceEditorContext()$path %>% str_split(pattern = "/") %>% unlist()
  document_name <- document_path[length(document_path)]
  sys_time <- Sys.time()
  if (grepl(Date, file) == T) {
    write(c(document_name, as.character(sys_time),"Output", correct_file), file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
    saveRDS(object, file)
  }
  else {
    print("Incorrect datetag!")
  }
}

#Read rds check through different versions
readRDS_proof <- function(path, pat) {
  full.path <- paste0(in.dir, path)
  document_path <- rstudioapi::getSourceEditorContext()$path %>% str_split(pattern = "/") %>% unlist()
  document_name <- document_path[length(document_path)]
  sys_time <- Sys.time()
  
  if(file.exists(full.path) == T){
    if(paste0(in.dir,path) != getwd()){
      setwd(paste0(in.dir, path))
  }
    file.list <- list.files(pattern = pat)
    correct_file <- str_sort(file.list, decreasing = T)[1]
    print(correct_file)
    write(c(document_name,as.character(sys_time) ,"Input", correct_file), file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
    readRDS(correct_file)
    
  }
  else{
  setwd(path)
  file.list <- list.files(pattern = pat)
  correct_file <- str_sort(file.list, decreasing = T)[1]
  print(correct_file)
  write(c(document_name, as.character(sys_time),"Input", correct_file),file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
  readRDS(correct_file)
  }
}
```

# Libraries
```{r libraries, warning=FALSE, message=FALSE}
# libraries:  
library(tidyverse)
library(ggrepel)
library(pheatmap)
library(rstatix)
library(stats)
library(tibble)
library(reshape2)
library(ggbeeswarm)
library(scales)
```

# Import data tables
```{r import, warning=FALSE, message=FALSE}

setwd(in.dir)
# Import TIDE results
# Fig S7A: Screen summary
screen_TIDE_summary <- read.csv("data/processed_data/TIDE_results/xv20240430_TIDE_screen_data_summary.csv")
# Fig S7I: Other RPE1 cells
othercells_TIDE_summary <- read.csv("data/processed_data/TIDE_results/xv20240430_TIDE_RPE1_cell_lines.csv")

#Import MMEJ:NHEJ balance ratio measurements per replicate
ddr.screen.all <- readRDS_proof( "data/processed_data/CCD_analysis","raw_data")

inhibtor.screen.all <- readRDS_proof("data/processed_data/CCD_inhibitors","repair_metrics")

inhibtor.screen.summary <- readRDS_proof("data/processed_data/","drug_CCD")

screen.summary <- readRDS_proof("data/processed_data/","Table_S7_DR")

revision_data_RPE1 <- readRDS_proof("/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_screen/","RPE1_control")
revision_data_PRO <- readRDS_proof("/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_screen/","PRO_control")
revision_data_DEF <- readRDS_proof("/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_screen/","DEF_control")

# heatmap chromatin order
heatmap.chromatin.order <- c("HDAC3","LMNB1","late_replicating","H3K9me2","H3K9me3","HDAC1","EZH2","H3K27me3","CTCF","SMC3","m5C","HDAC2","POL2","POL2AS2","H3K36me3","TTseq","Dam","H3K79me2","H2AFZ","DNAse","H3K4me3","H4K5acK8ac","H3K4me1","H3K27ac","H3K4me2")

clone5_bc_cluster <-  c("CATCCACCACACTTCA","ATATCGTTGCTGGAGA","GTACCTCTCGATAGTG","TCTTTTGAGGAGCTGA","CATTTCTGATCAATAA","TGGCCAATATTTGTCT","ACTGTCGAGTTGTCCG","GAGCGCGTCACCGGGT","GCGCACCCTTTAATTG","ACCCCTAAAGGCGCTG","CGGCCTGAAGGTCAGG","AGAAAATAATATGACG","TTGAACGCGGGCTCGG","CCGGGGACGTATGCAC","GCTAACATCACGAATC","ATACTATATTTAACGG","TATGGCTGTCGGGTAG","AGGGCGTAAAATATTT","TGTCCCTTAGTACTTT")

#Import delta log2 balance effects
PRO_log2 <-  readRDS("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_screen/xv20230915_PRO_all_mean_freq_0.0025.rds") %>% filter(is.finite(log2_bal))
K562_log2 <- readRDS("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/processed_data/CCD_analysis/xv20220819_screening_raw_data_repair_metrics.rds")
DEF_log2 <- readRDS("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_screen/xv20230911_DEF_all_mean_freq_0.0075_.rds") %>%
  filter(is.finite(log2_bal))

#Processed CCD data (4 cell lines)
PRO_CCD <- readRDS_proof("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/processed_data","PRO_CCD")
DEF_CCD <- readRDS_proof("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/processed_data","DEF_CCD")
K562_CCD_all <- readRDS_proof("~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/processed_data","Table_S7_DR") %>%
  mutate(cell_line = "K562") %>% 
  select(-screen_position,-DR_effect_type) %>% 
  dplyr::group_by(gene) %>%
  dplyr::mutate(pathway_signif = case_when(sum(CCD_synergy_score < 0) == 0 & sum(CCD_synergy_score > 0) != 0 ~ "NHEJ",
                                    sum(CCD_synergy_score > 0) == 0 & sum(CCD_synergy_score < 0) != 0 ~ "MMEJ",
                                    sum(CCD_synergy_score > 0) != 0 & sum(CCD_synergy_score < 0) != 0 ~ "both",
                                    T ~ "none"),
                n = 19,
                toxicity = "non_toxic")

proteins_gRNA <- readRDS_proof("/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/validation_screen","gRNA_gene") #gRNA gene conversion


K562_CCD_validation <- K562_CCD_all %>%
  distinct() %>%
  right_join(proteins_gRNA) %>%
  select(-gRNA)

#Bind all table together
cell_lines_CCD <- bind_rows(PRO_CCD, DEF_CCD,K562_CCD_validation)

genomewide_ranges <- readRDS("/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/data/revisions/pools_balance/xv202301009_RPE1_genomewide_ranges.rds")


#DNA repair protein KO CCD analysis
siRNA_CCDs <- readRDS_proof("data/processed_data/","S7_siRNA")
colnames(siRNA_CCDs)[1] <- "gene"
screen_CCDs <- readRDS_proof("data/processed_data/","S7_DR") %>% filter(gene %in% c("BRCA1","BRCA2","RAD51","POLQ"))

#Global effect
siRNA_delta <- readRDS_proof("data/processed_data/CCD_siRNA/","differentials_log2")
colnames(siRNA_delta)[1] <- "gene"
screen_delta <- readRDS_proof("data/processed_data/CCD_analysis","differentials_log2") %>% filter(gene %in% c("BRCA1","BRCA2","RAD51","POLQ"))

#Bind rows
all_CCDs <- bind_rows(siRNA_CCDs, screen_CCDs)
all_deltas <- bind_rows(siRNA_delta, screen_delta)

#Levels
siRNA_levels <- c("BRCA1","siBRCA1","RAD51","siRad51","BRCA2","siBRCA2","POLQ","siPolQ")


```

# Figure S7A
```{r, fig.width=3,fig.height=4, message=F,warning=F}
# Based on TIF
ggplot(screen_TIDE_summary,aes(replicate, editing_freq)) + 
  stat_summary(geom= "col") + 
  ggbeeswarm::geom_quasirandom() + 
  theme_bw() + facet_grid(~ transfection) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        panel.grid = element_blank()) + ylim(c(0,1))
```
(A) Total indel frequency in control locus (endogenous LBR gene) measured by TIDE 2 in each repli-cate. Total indel frequency is calculated as the sum of the percentage of all detectable indels by the TIDE algorithm. During the screening procedure, we measured editing frequency after the first transfec-tion (Day 5) and the second transfection (Day 9) in every plate in the screen (n = 6). In this plot only TIDE data of sufficient quality (TIDE R2 > 0.75) are included. Overall, TIDE estimated indel frequencies ranging from 0.37 to 0.71 for both transfections. 

# Figure S7B
## Data processing
```{r, warning=F, message=F}
#Merge transfection and nucleofection examples
editing_efficiency_WT <-  ddr.screen.all %>% filter(sample == "WT") %>% dplyr::select(barcode,well,plate, rep, freqCut) %>% mutate(delivery = "lipofection")
editing_efficiency_inhibitor <-  inhibtor.screen.all %>% filter(drug == "DMSO") %>% dplyr::select(barcode, rep = replicate, freqCut) %>% mutate(delivery = "nucleofection")

mergedt_K562_data <- editing_efficiency_WT %>% bind_rows(editing_efficiency_inhibitor)

#Calculate value per replicate
mean_per_replicate_TIF <- mergedt_K562_data %>%
  dplyr::group_by(barcode, delivery) %>%
  dplyr::summarise(mean_TIF = mean(freqCut))
 
#Reporters in transcribed regions 
transcr_reporters <- clone5_bc_cluster[c(12:19)]
editing_eff_transcr_reporters_inhibitor <- editing_efficiency_inhibitor %>% filter(barcode %in% transcr_reporters) 
```

## Plot
```{r, fig.width=2.5, fig.height=4, warning=F, message=F}
#Calculate mean per position
editing_efficiency_transcribed <- editing_efficiency_WT %>% filter(barcode %in% transcr_reporters) %>%
  dplyr::group_by(barcode, rep) %>%
  dplyr::summarise(TIF = mean(freqCut))

#Calculate mean per replication
mean_per_replicate <- editing_efficiency_transcribed %>%
  dplyr::group_by(rep) %>%
  dplyr::summarise(mean_TIF = mean(TIF))

#Plot editing frequency in transcribed reporters
#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231009_S7B_editing_frequency_screen_HTS.pdf", width = 2.5, height= 4)
ggplot(editing_efficiency_transcribed, aes(rep, TIF)) +
  stat_summary(geom = "col",fun.y = "mean") +
  geom_quasirandom() +
  ylab("Total indel frequency") +
  theme_bw()  +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank()) +
  ylim(c(0,1))
#dev.off()


```

(B) Complementary to TIDE, we calculated the mean total indel frequency in IPRs located in tran-scribed regions (n = 8) in K562 clone 5 for each replicate in mock transfected samples (n = 33). We chose IPRs in transcribed regions as they are most representative of the genes that we targeted in the KO screen. In this set-up, the total indel frequency is calculated as the frequency of reads with any insertion or deletion. By high-throughput sequencing, we estimated that the total editing frequency ranged from 0.59 to 0.7. Note that K562 cells are mostly triploid, and hence the percentage of cells with complete KO may be expected to be lower than these estimates. This is likely to account for the relatively small effect sizes observed in the K562 screen. To investigate the real extent of the effect-sizes, we performed the follow-up experiments in systems with higher penetrance. 


# Figure S7C

```{r, fig.width=3, fig.height=5}
#Create a tibble with pathway favor
favr_path <- tibble(gene = siRNA_levels, pathway = c("MMEJ","MMEJ","MMEJ","MMEJ","NHEJ","NHEJ","MMEJ","MMEJ"))

#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231009_S6A_siRNA_effect_size_difference.pdf", width = 3, height= 5)
ggplot(all_deltas %>% filter(gene != "siNT") %>% left_join(favr_path), aes(fct_relevel(gene, siRNA_levels),mean.log2foldchange)) +
  stat_summary(geom = "col", aes(fill = pathway)) +
  geom_quasirandom() +
  theme_bw() +
  geom_hline(yintercept = 0, linetype = 2) +
  theme(panel.grid = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "top") +
  scale_fill_manual(values = c("#8C510A","#01665E")) +
  ylab("∆log2MMEJ:NHEJ")
#dev.off()
```

(C-F) First, we re-analyzed previously published siRNA knockdown data (n = 3). (C) Comparison of ∆log2MMEJ:NHEJ scores of BRCA1, RAD51, BRCA2 and POLQ in the screen and after siRNA knock-downs from 3 (n = 19). In this experiment, the knock-down efficiency ranged from 50% to 80%.


# Figure S7D
```{r, fig.width=5, fig.height=4.5, warning=F, message=F}
#siRNA filter CCD plot
CCD_plot_siRNA <- all_CCDs %>% 
  filter(!is.na(chrom_feature) &
         !gene %in% c("BRCA1","siNT")) %>%
  mutate(target = case_when(gene == "siBRCA1" ~ "BRCA1",
                            gene == "siRad51" ~ "RAD51",
                            gene == "siPolQ" ~ "POLQ",
                            T ~ gene))

# Plot heatmap
#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231009_S6B_siRNA_CCD_effect.pdf", width = 5, height= 4.5)
ggplot(CCD_plot_siRNA) + 
  geom_tile(aes(fct_relevel(gene, siRNA_levels), fct_relevel(chrom_feature,heatmap.chromatin.order), fill = CCD_synergy_score)) +
  scale_fill_gradient2(low = "#8c510a" ,mid = "#f5f5f5", high = "#01665e", breaks = c(-0.1,0, 0.1), limits = c(-0.15,0.15), oob = squish)  + 
  coord_fixed(expand = F) + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title = element_blank(), legend.position = "top")
#dev.off()
```

(D) CCD patterns of RAD51, BRCA2 and POLQ in the screen and siRNA experiment after averaging all three replicates. CCD patterns of BRCA1 only in the siRNA experiment. Chromatin features are colored as in Fig. 1C. 

# Figure S7E-F
```{r, fig.width=4,fig.height=2, warning = F, message=F}
all_CCD_plot <- all_CCDs %>%
    mutate(target = case_when(gene == "siBRCA1" ~ "BRCA1",
                            gene == "siBRCA2" ~ "BRCA2",
                            gene == "siRad51" ~ "RAD51",
                            gene == "siPolQ" ~ "POLQ",
                            T ~ gene),
           technique = case_when(grepl("si", gene) ~ "siRNA",
                                 T ~ "screen"))

#Global effects
#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231009_S6C_CCD_effect_size.pdf", width = 4, height= 2.5)
ggplot(all_CCD_plot %>% select(technique, global_diff, target, gene) %>% distinct() %>% filter(gene != "siNT" & target != "BRCA1")) +
  geom_col(aes(technique, abs(global_diff))) +
  facet_wrap(~ target, ncol = 4) +
  theme_bw() +
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))  +
  ylab("Global ∆log2MMEJ:NHEJ")
#dev.off()

global_fc_siRNA <- all_CCD_plot %>% 
  select(technique, global_diff, target, gene) %>% 
  distinct() %>% 
  filter(gene != "siNT" & target != "BRCA1") %>%
  reshape2::dcast(target ~ technique, value.var = "global_diff") %>%
  mutate(fc = 2^(abs(siRNA) - abs(screen)))

#Fold changes CCD
#POLQ = 1.79
#BRCA2 = 1.3
#RAD51 = 3.54
#BRCA1 = 32.11 (didn't work in the screen)

#Global effects
max_CCD_effect <- all_CCD_plot %>% 
  select(technique, CCD_estim_diff, target, gene) %>% 
  distinct() %>% 
  filter(gene != "siNT") %>%
  dplyr::group_by(gene) %>%
  slice_max(abs(CCD_estim_diff))

#Calculate fold changes
fc_CCDs_siRNA <- max_CCD_effect %>%
  reshape2::dcast(target ~ technique, value.var = "CCD_estim_diff") %>%
  mutate(fc = 2^(abs(siRNA) - abs(screen)))


#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231009_S6D_global_effect_size.pdf", width = 4, height= 2.5)
ggplot(max_CCD_effect %>% filter(target != "BRCA1")) +
  geom_col(aes(technique, abs(CCD_estim_diff))) +
  facet_wrap(~ target, ncol = 4) +
  theme_bw() +
  theme(panel.grid = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab("Max. CCD ∆log2MMEJ:NHEJ")
#dev.off()
#Fold changes CCD
#POLQ = 1.16
#BRCA2 = 1.91
#RAD51 = 1.28
```
(E-F) Knock-down of BRCA2, RAD51 and POLQ yield larger (E) global (∆Global) and (F) CCD (Max. ∆CCD) log2MMEJ:NHEJ scores than in the screen. We calculated ∆Global and ∆CCD as described in Figure S6. 


# Figure S7G-H
## Data processing
```{r, warning=F, message=F}
#Effect sizes of drug treatment (combinde data tables)
KO_data_K562 <- screen.summary %>% select(gene, CCD_synergy_score, CCD_estim_diff,global_diff, chrom_feature) %>% mutate(perturbation = "Screen") %>% filter(gene %in% c("ATM","PRKDC"))
drug_data_K562 <- inhibtor.screen.summary %>% select(gene = drug, CCD_synergy_score, CCD_estim_diff,global_diff, chrom_feature) %>% mutate(perturbation = "Inhibitor")

combined_data <- bind_rows(KO_data_K562, drug_data_K562) %>%
  mutate(gene = case_when(grepl("ATM", gene) ~ "ATM",
                          gene == "DNAPKi" ~ "PRKDC",
                          T ~ gene)) %>% 
  mutate(effect_size = abs(CCD_estim_diff)) %>%
  dplyr::group_by(gene,perturbation) %>%
  slice_max(effect_size)

```

## Plot Figure S7G
```{r, fig.width=5,fig.height=3, warning=F, message=F}
#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231009_inhibitor_effect_size_global.pdf", width = 5, height= 3)
ggplot(combined_data) +
  geom_col(aes(fct_relevel(perturbation,c("Screen","Inhibitor")),abs(global_diff))) +
  facet_wrap(~ gene) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank()) +
  ylab("Global ∆log2MMEJ:NHEJ") + ylim(c(0,7))
#dev.off()

#fold change calculation
#ATM
ATM_global_fc <- 2^(1.86 - 0.118) #3.3 fold-change
DNAPK_global_fc <- 2^(6.21 - 0.086) #69 fold-change 
```

## Plot Figure S7D
```{r, fig.width=5,fig.height=3, warning=F, message=F}
#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231009_inhibitor_effect_size_CCD.pdf", width = 5, height= 3)
ggplot(combined_data) +
  geom_col(aes(fct_relevel(perturbation,c("Screen","Inhibitor")),abs(CCD_estim_diff))) +
  facet_wrap(~ gene) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank()) +
  ylab("Max. CCD ∆log2MMEJ:NHEJ")
#dev.off()

#fold change calculation
#ATM
ATM_CCD_fc <- 2^(1.16 - 0.18) #1.97 fold-change
DNAPK_CCD_fc <- 2^(0.901 -0.11) #1.73 fold-change 
```
(G-H) Second, we treated K562 clone 5 cells with small molecule inhibitors targeting ATM and PRKDC. Inhibitor treatments lead to (G) global (∆Global) and (H) CCD (Max. ∆CCD) effect sizes larger than KO of the inhibited protein in the screen. We calculated ∆Global and ∆CCD as described in Figure S6.

 
# Figure S7I
```{r, fig.width=3,fig.height=5, warning=F, message=F}
cell_levels <- c("K562","DEF","PRO")
# Based on TIF
#pdf(file = "/Users/x.vergara.ucin/surfdrive/Shared/surfdrive_shared_steensel/writing_spree/draft_DSB_CRISPR_screen/short_version/Figures/source_figures/Supplementary_figures/Figure_S6/xv20231117_LBR2_TIDE_other_cell_lines.pdf", width = 3, height = 5)
ggplot(othercells_TIDE_summary %>% filter(cell_line != "RPE1"),aes(fct_relevel(cell_line, c("K562","PRO","DEF")), editing_freq, fill = cell_line)) + 
  stat_summary(geom= "col", position = position_dodge2(preserve = "single")) + 
  ggbeeswarm::geom_quasirandom() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), legend.position = "none", panel.grid = element_blank()) + ylim(c(0,1)) + 
  labs(fill = "Condition") +
  xlab("Cell line") + 
  scale_fill_manual(values = c("#9f86bf","grey70","#95c482")) +
  ylab("Total editing frequency (by TIDE)")
#dev.off()

#Test to compare different editing frequencies
t_test_TIF <- othercells_TIDE_summary %>%
  filter(cell_line != "RPE1") %>%
  ungroup() %>%
  rstatix::t_test(editing_freq ~ cell_line, ref.group = "K562", alternative = "less") %>%
  mutate(fdr = p.adjust(p, method = "BH"))
```
(I-J) We measured the total indel frequency in RPE-1 cells compared to K562 (I) by TIDE (K562 n = 6, RPE-1 p53KO n = 5 and RPE-1 p53/BRCA1dKO n = 7) 


# Figure S7J
## Data processing
```{r, fig.width=3, fig.height=4, warning=F, message=F}
#Merge transfection and nucleofection examples
editing_efficiency_PRO <-  revision_data_PRO %>% mutate(TIF = 1 - (wt_0/total_IPR_reads)) %>% filter(gRNA == "halfLBR2") %>% select(barcode, cell_line,rep = bio_rep, TIF)
editing_efficiency_DEF <-  revision_data_DEF %>% mutate(TIF = 1 - (wt_0/total_IPR_reads)) %>% filter(gRNA == "LBR2") %>% select(barcode, cell_line,rep = bio_rep, TIF)

#K562 data
editing_efficiency_K562 <- mergedt_K562_data %>% 
  filter(delivery == "lipofection" & barcode %in% transcr_reporters) %>%
  ungroup() %>%
  dplyr::group_by(rep) %>%
  dplyr::summarise(mean_TIF = mean(freqCut)) %>%
  mutate(cell_line = "K562")



#Merge RPE1 data
merge_RPE1_data_TIF <- bind_rows(editing_efficiency_PRO, editing_efficiency_DEF) %>% 
  dplyr::group_by(rep,cell_line) %>% 
  dplyr::summarise(mean_TIF = mean(TIF)) %>%
  bind_rows(editing_efficiency_K562)

#Calcualte TIF
t_test_TIF <- merge_RPE1_data_TIF %>%
  ungroup() %>%
  t_test(mean_TIF ~ cell_line, ref.group = "K562", alternative = "less")
```

## Plot
```{r, warning=F, message=F}
#Plot as beeswarm plot
#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231009_TIF_validation_experiment.pdf", width = 3, height= 5)
ggplot(merge_RPE1_data_TIF,aes(fct_relevel(cell_line, c("K562","PRO","DEF")), mean_TIF)) +
  stat_summary(geom = "col",fun.y = "mean", aes(fill = cell_line)) +
  ggbeeswarm::geom_quasirandom() +
  ylab("Total indel frequency") +
  theme_bw()  +
  scale_fill_manual(values = c("#9f86bf","grey70","#95c482")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank(),
        panel.grid = element_blank())
#dev.off()
```
and (J) high-throughput sequencing (n = 3). With both techniques, we detected a nearly complete editing frequency in the LBR control locus in RPE-1 cells. Significance was assessed by two-sided Wilcoxon test with correction for multiple testing by the Benjamini-Hochberg method. 

# Figure S7K
```{r, fig.width=3, fig.height=4, warning=F, message=F}
global_effects_plot <- cell_lines_CCD %>% select(gene,global_diff, cell_line) %>% distinct() %>% na.omit()
#Overall ∆log2 balance global changes
#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231117_global_effect.pdf", width = 3, height= 5)
ggplot(global_effects_plot %>% filter(cell_line != "RPE1"), aes(fct_relevel(cell_line, c("K562","PRO","DEF")), abs(global_diff))) +
  stat_summary(geom = "col",fun.y = "mean", aes(fill = cell_line)) +
  geom_quasirandom() +
  ylab("abs(Global ∆log2 MMEJ::NHEJ)") +
  theme_bw()  +
  scale_fill_manual(values = c("#9f86bf","grey70","#95c482")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank(),
        panel.grid = element_blank())
dev.off()

#Perform test (differences between cell lines?)

global_effect_test <- map_dfr(c("DEF","PRO"), function(x) {
  cell_line1 <- cell_lines_CCD %>% filter(cell_line == "K562") %>% select(cell_line, gene, global_diff) %>% distinct()
  cell_line2 <- cell_lines_CCD %>% filter(cell_line == x) %>% select(cell_line,gene, global_diff) %>% distinct()
  proteins <- intersect(cell_line1$gene, cell_line2$gene)
  cell_lines <- bind_rows(cell_line1, cell_line2) %>% filter(gene %in% proteins) %>% distinct() %>% mutate(abs_global = abs(global_diff))
  cell_lines %>% ungroup() %>% wilcox_test(abs_global ~ cell_line, paired = T, ref.group = "K562", alternative = "less")
  }) %>% mutate(fdr = p.adjust(p, method = "BH"))

```


# Figure S7L
```{r, fig.width=3, fig.height=4, warning=F, message=F}
#Create data frame with only matching pairs
strongest_CCD_proteins <- cell_lines_CCD %>% 
  filter(CCD_synergy_score != 0 & CCD_model_p_adj < 0.25) %>%
  mutate(effect_CCD = abs(CCD_estim_diff)) %>%
  dplyr::group_by(gene,cell_line) %>%
  dplyr::slice_max(effect_CCD)

#pdf(file = "~/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/export/xv20220117_paper_figure_short/supplementary_figures/figure_S6_editing_efficiency/xv20231017_CCD_effect.pdf", width = 3, height= 5)
ggplot(strongest_CCD_proteins, aes(fct_relevel(cell_line, c("K562","PRO","DEF")), effect_CCD)) +
  stat_summary(geom = "col",fun.y = "mean", aes(fill = cell_line)) +
  ggbeeswarm::geom_quasirandom() +
  ylab("abs(Max. CCD ∆log2MMEJ:NHEJ)") +
  theme_bw()  +
  scale_fill_manual(values = c("#9f86bf","grey70","#95c482")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank(),
        panel.grid = element_blank())
#dev.off()


CCD_greater_test <- map_dfr(c("DEF","PRO"), function(x) {
  cell_line1 <- strongest_CCD_proteins %>% filter(cell_line == "K562") %>% select(cell_line, gene, effect_CCD) 
  cell_line2 <- strongest_CCD_proteins %>% filter(cell_line == x) %>% select(cell_line, gene, effect_CCD)
  proteins <- intersect(cell_line1$gene, cell_line2$gene)
  cell_lines <- bind_rows(cell_line1, cell_line2) %>% filter(gene %in% proteins) %>% distinct()
  cell_lines %>% ungroup() %>% wilcox_test(effect_CCD ~ cell_line, paired = T, ref.group = "K562", alternative = "less")
  }) %>% mutate(fdr = p.adjust(p, method = "fdr"))

```

K-L) ∆log2MMEJ:NHEJ in RPE-1 cells for the 20 targeted proteins, show higher (K) global (∆Global) and (L) CCD (Max. ∆CCD) effect sizes than in K562 cells. We calculated ∆Global and ∆CCD as described in Figure S6. Significance of the observed effects was measured by Wilcoxon test with correction for mul-tiple testing by Benjamini-Hochberg method (n = 20). We conclude that the K562 screen underesti-mated the effect sizes of both the global and CCD effects, most likely due to low KO efficiencies.
