---
title: "xv20220329_FA_SCC_mutations_analysis"
author: "x.vergara"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

In this document, I will go over different calculations of epistasis.4 different workflows:
First decision is whether to use log2 MMEJscore or MMEJscore
Second decision is comparing slopes or statistic differences between WT vs. KO slopes

```{r}
knitr::opts_chunk$set(root.dir = '/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/rmarkdown/xv20210920_paper_scripts/short_version/knit_html/')

StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

#Change in.dir accordingly
in.dir <- "/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/"

#Check if output contain datetag
saveRDS_proof <- function(object, file) {
  if (grepl(Date, file) == T) {
    saveRDS(object, file)
  }
  else {
    print("Incorrect datetag!")
  }
}

#Read rds check through different versions
readRDS_proof <- function(path, pat) {
  full.path <- paste0(in.dir, path)
  document_path <- rstudioapi::getSourceEditorContext()$path %>% str_split(pattern = "/") %>% unlist()
  document_name <- document_path[length(document_path)]
  sys_time <- Sys.time()
  
  if(file.exists(full.path) == T){
    if(paste0(in.dir,path) != getwd()){
      setwd(paste0(in.dir, path))
  }
    file.list <- list.files(pattern = pat)
    correct_file <- str_sort(file.list, decreasing = T)[1]
    print(correct_file)
    write(c(document_name,as.character(sys_time) ,"Input", correct_file), file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
    readRDS(correct_file)
    
  }
  else{
  setwd(path)
  file.list <- list.files(pattern = pat)
  correct_file <- str_sort(file.list, decreasing = T)[1]
  print(correct_file)
  write(c(document_name, as.character(sys_time),"Input", correct_file),file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
  readRDS(correct_file)
  }
}

```

## Libraries:

```{r libraries}
library(readxl)
library(tidyverse)
library(GenomicRanges)
library(rtracklayer)
library(vcfR)
```

## Import data

```{r pressure, echo=FALSE}
setwd(in.dir)

#Import sporadic mutations
sporadic_mutation_dt <- read_xlsx("import/Webster_et_al_2021_FA_tumours/xv20220329_T5_FA_SCC_SV.xlsx", col_names = T, sheet = 2, skip = 1)
#New analysis
sporadic_mutation_dt_bis <- read_delim("import/xv20220727_PCAWG_tumor_data/SV_Indels_PCAWG/HNSC/BRASS/summary_shortened.txt")

#Import sporadic mutations
BRCA2_mutation_dt <- read_xlsx("import/Webster_et_al_2021_FA_tumours/xv20220329_T5_FA_SCC_SV.xlsx", col_names = T, sheet = 3, skip = 1)
#New analysis
BRCA2_mutation_dt_bis <- read_delim("import/xv20220727_PCAWG_tumor_data/SV_Indels_PCAWG/BRCA2/BRASS/summary_shortened.txt")

```

#Import LADs data
```{r}
setwd(in.dir)
# LAD domains
LAD_atlas <- as.data.frame(read.table("import/xv20220329_LAD_atlas_OSF.bed.gz",header = FALSE, sep="\t",stringsAsFactors=FALSE, quote=""))
colnames(LAD_atlas) <- c("chr","start","end","length","strand","LAD_class")

#LAD_length plot control
LAD_length_tibble <- LAD_atlas %>% mutate(LAD_length = end - start)

```

#Curate data tables: What are the features that I am interested (only select deletions)
```{r}
#Select only deletions (in all four files)
sporadic_mutation_del_dt <- sporadic_mutation_dt %>% filter(svclass == "deletion" & !`micro-homology` %in% c(".","_") & `Brass Notation` != "_")
sporadic_mutation_del_dt_bis <- sporadic_mutation_dt_bis %>% filter(svclass == "deletion" & !"micro-homology" %in% c(".","_") & `Brass Notation` != "_")
#Check by inner join
common_part_sporadic <- inner_join(sporadic_mutation_del_dt,sporadic_mutation_del_dt_bis, by = c("start1","end1","chr2","start2","end2"))
#Numeric brass score
sporadic_brass_score <- sporadic_mutation_del_dt$`Brass Notation` %>% str_extract("(?<=score ).*(?=\\))") %>% as.numeric()
sporadic_brass_score_bis <- sporadic_mutation_del_dt_bis$`Brass Notation` %>% str_extract("(?<=score ).*(?=\\))") %>% as.numeric()

#BRCA2 data ()
BRCA2_mutation_del_dt <- BRCA2_mutation_dt %>% filter(svclass == "deletion"  & !"micro-homology" %in% c(".","_") & `Brass Notation` != "_")
BRCA2_mutation_del_dt_bis <- BRCA2_mutation_dt_bis %>% filter(svclass == "deletion"  & !"micro-homology" %in% c(".","_") & `Brass Notation` != "_")
#Check by inner join
common_part_BRCA2 <- inner_join(BRCA2_mutation_del_dt,BRCA2_mutation_del_dt_bis, by = c("start1","end1","chr2","start2","end2"))
#numeric brass score
BRCA2_brass_score <- BRCA2_mutation_del_dt$`Brass Notation` %>% str_extract("(?<=score ).*(?=\\))") %>% as.numeric()
BRCA2_brass_score_bis <- BRCA2_mutation_del_dt_bis$`Brass Notation` %>% str_extract("(?<=score ).*(?=\\))") %>% as.numeric()
```

#Map SV for BRCA2 and sporadic tumors
```{r}
#Create function to call unique LADs
LAD_atlas_ranges <- makeGRangesFromDataFrame(LAD_atlas, keep.extra.columns = T)
unique_LADs_call <- function(x) {
  mutations_GR <- makeGRangesFromDataFrame(x, keep.extra.columns = T)
  LAD_overlaps <- findOverlaps(LAD_atlas_ranges,mutations_GR)
  LAD_overlaps_dt <- tibble(x[subjectHits(LAD_overlaps),], LAD_status =  LAD_atlas[queryHits(LAD_overlaps),6])
  LAD_counts <- LAD_overlaps_dt %>% dplyr::group_by(sample, id) %>% dplyr::summarize(count = n())
  single_LADs_iLADs <-LAD_counts %>% filter(count == 1) %>% select(sample, id)
  LAD_overlaps_dt %>% right_join(single_LADs_iLADs)
}

#Prepare the data table for mapping
mutation_location_export_spo_SV <- sporadic_mutation_del_dt %>% mutate(mut_id = paste(sample, `id/name`, sep = "_"), chr = paste0("chr", `# chr1`)) %>% select(chr, start = "start1", end = "start2", mut_id, sample, id = `id/name`)
mutation_location_export_spo_SV_bis <- sporadic_mutation_del_dt_bis %>% mutate(mut_id = paste(sample, `id/name`, sep = "_"), chr = paste0("chr", `chr1`)) %>% select(chr, start = "start1", end = "start2", mut_id, sample, id = `id/name`)
mutation_location_export_BRCA2_SV <- BRCA2_mutation_del_dt %>% mutate(mut_id = paste(sample, `id/name`, sep = "_"), chr = paste0("chr", `# chr1`)) %>% select(chr, start = "start1", end = "start2", mut_id, sample, id = `id/name`)
mutation_location_export_BRCA2_SV_bis <- BRCA2_mutation_del_dt_bis %>% mutate(mut_id = paste(sample, `id/name`, sep = "_"), chr = paste0("chr", `chr1`)) %>% select(chr, start = "start1", end = "start2", mut_id, sample, id = `id/name`)

#Run function
spo_SV_LAD_maps <- unique_LADs_call(mutation_location_export_spo_SV) #526 total MH del
spo_SV_LAD_maps_bis <- unique_LADs_call(mutation_location_export_spo_SV_bis) #993 total MH del
BRCA2_SV_LAD_maps <- unique_LADs_call(mutation_location_export_BRCA2_SV) #2022 total MH del
BRCA2_SV_LAD_maps_bis <- unique_LADs_call(mutation_location_export_BRCA2_SV_bis) #2281 total MH del

#Calculate mean per sample
spo_SV_LAD_maps_tumor <- spo_SV_LAD_maps %>% dplyr::group_by(sample, grepl("i",LAD_status)) %>% dplyr::summarise(mut_n = n())
spo_SV_LAD_maps_tumor_bis <- spo_SV_LAD_maps_bis %>% dplyr::group_by(sample, grepl("i",LAD_status)) %>% dplyr::summarise(mut_n = n())
BRCA2_SV_LAD_maps_tumor <- BRCA2_SV_LAD_maps %>% dplyr::group_by(sample, grepl("i",LAD_status)) %>% dplyr::summarise(mut_n = n())
BRCA2_SV_LAD_maps_tumor_bis <- BRCA2_SV_LAD_maps_bis %>% dplyr::group_by(sample, grepl("i",LAD_status)) %>% dplyr::summarise(mut_n = n())
```

#Import all vcf files and make a single data table with them (All indels)
```{r}
#setwd
setwd(paste0(in.dir, "import/xv20220727_PCAWG_tumor_data/SV_Indels_PCAWG"))

#list.all files
BRCA2_pindel_files <- list.files(path = "BRCA2/PINDEL",pattern = ".vcf", full.names = T)
control_pindel_files <- list.files(path = "HNSC/PINDEL",pattern = ".vcf", full.names = T)

#Function to import and clean data for all tumours
process_vcf <- function(x) {
  vcf_file <- read.vcfR(x, verbose = F)
  vcf_remove <- vcf_file@fix %>% as_tibble() %>% separate(INFO, sep = ";", into = c("PC","RS","RE","LEN","S1","S2","REP")) %>% mutate_all(.funs = gsub, pattern = ".*=", replace = "") %>% mutate(sampleID = x)
}

#All data together for BRCA2 and controls
PINDEL_call_BRCA2 <- map_dfr(BRCA2_pindel_files, process_vcf)
PINDEL_call_HNSC <- map_dfr(control_pindel_files, process_vcf)

#Change length column
PINDEL_call_BRCA2$LEN <- as.numeric(PINDEL_call_BRCA2$LEN)
PINDEL_call_HNSC$LEN <- as.numeric(PINDEL_call_HNSC$LEN)
PINDEL_call_BRCA2$QUAL <- as.numeric(PINDEL_call_BRCA2$QUAL)
PINDEL_call_HNSC$QUAL <- as.numeric(PINDEL_call_HNSC$QUAL)
PINDEL_call_BRCA2$POS <- as.numeric(PINDEL_call_BRCA2$POS)
PINDEL_call_HNSC$POS <- as.numeric(PINDEL_call_HNSC$POS)

#Filter deletions only
PINDEL_call_BRCA2_del <- filter(PINDEL_call_BRCA2 , PC == "D" & QUAL > 500 & LEN > 20 & REP == "1") %>% rowwise() %>% mutate(ALT_l = length(ALT))
PINDEL_call_HNSC_del <- filter(PINDEL_call_HNSC , PC == "D" & QUAL > 500 & LEN > 20 & REP == "1") %>% rowwise() %>% mutate(ALT_l = length(ALT))

```

#Map Indels for BRCA2 and sporadic tumors
```{r}
#Create function to call unique LADs
LAD_atlas_ranges <- makeGRangesFromDataFrame(LAD_atlas, keep.extra.columns = T)
unique_LADs_call <- function(x) {
  mutations_GR <- makeGRangesFromDataFrame(x, keep.extra.columns = T)
  LAD_overlaps <- findOverlaps(LAD_atlas_ranges,mutations_GR)
  LAD_overlaps_dt <- tibble(x[subjectHits(LAD_overlaps),], LAD_status =  LAD_atlas[queryHits(LAD_overlaps),6])
  LAD_counts <- LAD_overlaps_dt %>% dplyr::group_by(sample, id) %>% dplyr::summarize(count = n())
  single_LADs_iLADs <-LAD_counts %>% filter(count == 1) %>% select(sample, id)
  LAD_overlaps_dt %>% right_join(single_LADs_iLADs)
}

#Prepare the data table for mapping
mutation_location_export_spo_indel <- PINDEL_call_HNSC_del %>% rowwise() %>% mutate(mut_id = paste(sampleID, ID, sep = "_"), chr = paste0("chr", CHROM), end = POS + LEN) %>% select(chr, start = "POS", end, mut_id, sample = "sampleID", id = "ID")

mutation_location_export_BRCA2_indel <- PINDEL_call_BRCA2_del %>% rowwise() %>% mutate(mut_id = paste(sampleID, ID, sep = "_"), chr = paste0("chr", CHROM), end = POS + LEN) %>% select(chr, start = "POS", end, mut_id, sample = "sampleID", id = "ID")

#Run function
spo_indel_LAD_maps <- unique_LADs_call(mutation_location_export_spo_indel) #1062 total MH del
BRCA2_indel_LAD_maps <- unique_LADs_call(mutation_location_export_BRCA2_indel) #6068 total MH del

#Calculate mean per sample
spo_indel_LAD_maps_tumor <- spo_indel_LAD_maps %>% dplyr::group_by(sample, LAD_status) %>% dplyr::summarise(mut_n = n())
BRCA2_indel_LAD_maps_tumor <- BRCA2_indel_LAD_maps %>% dplyr::group_by(sample, LAD_status) %>% dplyr::summarise(mut_n = n())
```


```{r}
#Prepare the data table for export
mutation_location_export <- all_tumour_deletions_clasified %>% mutate(mut_id = paste(sample, id, sep = "_"), chr = paste0("chr", chr1)) %>% select(chr, start = "start1", end = "start2", strand = "strand1" ,mut_id, del_class, del_length)
mutation_location_GR <- makeGRangesFromDataFrame(mutation_location_export, keep.extra.columns = T)

#Crate ranges with LADS
LAD_atlas_ranges <- makeGRangesFromDataFrame(LAD_atlas, keep.extra.columns = T)
LAD_overlaps <- findOverlaps(LAD_atlas_ranges,mutation_location_GR)
LAD_overlaps_dt <- tibble(mutation_location_export[subjectHits(LAD_overlaps),], LAD_status =  LAD_atlas[queryHits(LAD_overlaps),6])

#Put both together
mut_data_LADs <- LAD_overlaps_dt %>% mutate(type = case_when(grepl("BRCA2", mut_id) ~ "BRCA2mut", grepl("BRCA1", mut_id) ~ "BRCA1mut",grepl("HNSCC", mut_id) ~ "spo", grepl("F",mut_id) ~ "FA"))  %>% separate(mut_id, sep = "_(?=[1-9])", into = c("sample","id"))
mut_data_LADs$id  <- as.double(mut_data_LADs$id)

# Some deletions are very big and span over LADs and iLADs (filter only deletions that occur within the same iLAD or LAD)
LAD_counts <- mut_data_LADs %>% dplyr::group_by(sample, id) %>% dplyr::summarize(count = n())
single_LADs_iLADs <-LAD_counts %>% filter(count == 1) %>% select(sample, id)

#Filter only FANCA tumours
sim.chrom.status.cd <- mut_data_LADs %>% right_join(single_LADs_iLADs) %>% left_join(FA_tumour_type_cg %>% distinct()) %>% mutate(type_cd = paste(type, CD, sep = "_")) %>% filter(type != "FA" | type_cd == "FA_A")

#Call HPV status of tumors
sim.chrom.status.cd.hpv <- sim.chrom.status.cd %>% rowwise() %>% mutate(HPV = grepl("HPVPos", sample))
```
