---
title: "xv20230817_mutations_analysis"
author: "x.vergara"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

I will explore the new data that Mathijs sent me, and figure it out if we can use it in the paper.

```{r}
knitr::opts_chunk$set(root.dir = '/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/rmarkdown/xv20210920_paper_scripts/short_version/knit_html/')

StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

#Change in.dir accordingly
in.dir <- "/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/import/xv20231017_indel_data_mathijs_new/BOOTSTRAP"

#Check if output contain datetag
saveRDS_proof <- function(object, file) {
  if (grepl(Date, file) == T) {
    saveRDS(object, file)
  }
  else {
    print("Incorrect datetag!")
  }
}

#Read rds check through different versions
readRDS_proof <- function(path, pat) {
  full.path <- paste0(in.dir, path)
  document_path <- rstudioapi::getSourceEditorContext()$path %>% str_split(pattern = "/") %>% unlist()
  document_name <- document_path[length(document_path)]
  sys_time <- Sys.time()
  
  if(file.exists(full.path) == T){
    if(paste0(in.dir,path) != getwd()){
      setwd(paste0(in.dir, path))
  }
    file.list <- list.files(pattern = pat)
    correct_file <- str_sort(file.list, decreasing = T)[1]
    print(correct_file)
    write(c(document_name,as.character(sys_time) ,"Input", correct_file), file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
    readRDS(correct_file)
    
  }
  else{
  setwd(path)
  file.list <- list.files(pattern = pat)
  correct_file <- str_sort(file.list, decreasing = T)[1]
  print(correct_file)
  write(c(document_name, as.character(sys_time),"Input", correct_file),file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
  readRDS(correct_file)
  }
}

```

## Libraries:
```{r libraries}
library(readxl)
library(tidyverse)
library(GenomicRanges)
library(rtracklayer)
library(ggbeeswarm)
```

#Import indel data
```{r, warning=F, message=F}
setwd(in.dir)

#Function to import all the data
#List of all driver mutations
driver_mutations_indels <- list.files(path = "/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/import/xv20231017_indel_data_mathijs_new/BOOTSTRAP", full.names = T)

control_indel_file_names <- map_dfr(driver_mutations_indels, function(y) {
  control_directory <- paste0(y,"/controls_indel")
  #Import all data for a single driver mutation
tumor_file_names <- list.files(path = control_directory, full.names = T)
  mutation_files <- map_dfr(tumor_file_names,function(x) {
    tibble(file_names = list.files(path = x, full.names = T))
  })
  })

#list of files
control_indel_data <- map_dfr(control_indel_file_names$file_names, function(x) {
  read.delim(x) %>%
    mutate(file = str_extract(x, "(?<=BOOTSTRAP/).*LAD")) %>%
    separate(file, into = c("driver",NA,"project","chromatin"), sep = "/")
}) %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>%
  reshape2::dcast(TYPE+driver+project~chromatin+variable, fun.aggregate = sum)  %>% 
  mutate(sample = "control")


#Load samples
indel_file_names <- map_dfr(driver_mutations_indels, function(y) {
  #Import all data for a single driver mutation
tumor_file_names <- list.files(path = y, full.names = T)
  mutation_files <- map_dfr(tumor_file_names,function(x) {
    tibble(file_names = list.files(path = x, full.names = T))
  })
  }) %>% filter(grepl(".txt", file_names))

#list of files
mutant_indel_data <- map_dfr(indel_file_names$file_names, function(x) {
  read.delim(`x`) %>%
    mutate(file = str_extract(x, "(?<=BOOTSTRAP/).*LAD")) %>%
    separate(file, into = c("driver","project","chromatin"), sep = "/")
}) %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>%
  reshape2::dcast(TYPE+driver+project~chromatin+variable, fun.aggregate = sum)  %>% 
  mutate(sample = "mutant")

#combine control and mutant data
all_indel_data_together <- control_indel_data %>% bind_rows(mutant_indel_data) %>% filter(!TYPE %in% multiple_drivers_data$TYPE)

```




#Load tumor info data
```{r}
#Import qualitative data
pcaw_TCGA_donor <- read_delim("/DATA/projects/DSBrepair/data/xv20230818_PCAWG_data_revision/DATA_PCAWG/INFO/pcawg_TCGA_to_donorID.txt")
selection <- read_delim("/DATA/projects/DSBrepair/data/xv20230818_PCAWG_data_revision/DATA_PCAWG/INFO/selection.txt")

#Combine them
tumor_info <- left_join(pcaw_TCGA_donor,selection, c("TCGA_ID" = "sample_id"))

#Multiple driver genes
multiple_drivers_data <- mutant_indel_data %>% 
  select(TYPE, project,driver) %>% 
  distinct() %>%
  dplyr::group_by(TYPE, project) %>%
  dplyr::summarise(counts = n()) %>%
  filter(counts > 1)


```

#Number of mutations per category
```{r, warning=F, message=F}
#Combine mutant and control
all_indel_data <- mutant_indel_data %>%  filter(!TYPE %in% multiple_drivers_data$TYPE)

#Per sample
lowest_indel_count_sample <- all_indel_data %>% 
  rowwise() %>%
  mutate(lowest_indel_count = min(CILAD_NHEJ,CILAD_MMEJ,CLAD_NHEJ,CLAD_MMEJ))
#Summary table
distribution_per_sample <- tibble(class = "patient",
                                  perc0.05 = quantile(lowest_indel_count_sample$lowest_indel_count, 0.05),
                                  perc0.25 = quantile(lowest_indel_count_sample$lowest_indel_count, 0.25),
                                  perc0.5 = quantile(lowest_indel_count_sample$lowest_indel_count, 0.5),
                                  perc0.75 = quantile(lowest_indel_count_sample$lowest_indel_count, 0.75),
                                  perc0.95 = quantile(lowest_indel_count_sample$lowest_indel_count, 0.95))

#Per project
lowest_indel_count_projects <- all_indel_data %>% 
  dplyr::group_by(driver,project) %>%
  dplyr::summarise(CILAD_NHEJ = sum(CILAD_NHEJ),
                   CILAD_MMEJ = sum(CILAD_MMEJ),
                   CLAD_NHEJ = sum(CLAD_NHEJ),
                   CLAD_MMEJ = sum(CLAD_MMEJ)) %>%
  rowwise() %>%
  mutate(lowest_indel_count = min(CILAD_NHEJ,CILAD_MMEJ,CLAD_NHEJ,CLAD_MMEJ))
#Summary table
distribution_per_project <- tibble(class = "tumor_type",
                                  perc0.05 = quantile(lowest_indel_count_projects$lowest_indel_count, 0.05),
                                  perc0.25 = quantile(lowest_indel_count_projects$lowest_indel_count, 0.25),
                                  perc0.5 = quantile(lowest_indel_count_projects$lowest_indel_count, 0.5),
                                  perc0.75 = quantile(lowest_indel_count_projects$lowest_indel_count, 0.75),
                                  perc0.95 = quantile(lowest_indel_count_projects$lowest_indel_count, 0.95))

#Per most common tumor types (more than 5 tumors)
lowest_indel_count_common_tumors <- all_indel_data %>% 
  separate(project, into =c("tumor", "country")) %>%
  right_join(tumor_types_cases) %>%
  dplyr::group_by(driver,tumor) %>%
  dplyr::summarise(CILAD_NHEJ = sum(CILAD_NHEJ),
                   CILAD_MMEJ = sum(CILAD_MMEJ),
                   CLAD_NHEJ = sum(CLAD_NHEJ),
                   CLAD_MMEJ = sum(CLAD_MMEJ)) %>%
  rowwise() %>%
  mutate(lowest_indel_count = min(CILAD_NHEJ,CILAD_MMEJ,CLAD_NHEJ,CLAD_MMEJ))
#Summary table
distribution_per_common_tumor <- tibble(class = "common_tumor",
                                  perc0.05 = quantile(lowest_indel_count_common_tumors$lowest_indel_count, 0.05),
                                  perc0.25 = quantile(lowest_indel_count_common_tumors$lowest_indel_count, 0.25),
                                  perc0.5 = quantile(lowest_indel_count_common_tumors$lowest_indel_count, 0.5),
                                  perc0.75 = quantile(lowest_indel_count_common_tumors$lowest_indel_count, 0.75),
                                  perc0.95 = quantile(lowest_indel_count_common_tumors$lowest_indel_count, 0.95))

#Per driver
lowest_indel_count_driver <- all_indel_data %>% 
  dplyr::group_by(driver) %>%
  dplyr::summarise(CILAD_NHEJ = sum(CILAD_NHEJ),
                   CILAD_MMEJ = sum(CILAD_MMEJ),
                   CLAD_NHEJ = sum(CLAD_NHEJ),
                   CLAD_MMEJ = sum(CLAD_MMEJ)) %>%
  rowwise() %>%
  mutate(lowest_indel_count = min(CILAD_NHEJ,CILAD_MMEJ,CLAD_NHEJ,CLAD_MMEJ))
#Summary table
distribution_per_driver <- tibble(class = "gene",
                                  perc0.05 = quantile(lowest_indel_count_driver$lowest_indel_count, 0.05),
                                  perc0.25 = quantile(lowest_indel_count_driver$lowest_indel_count, 0.25),
                                  perc0.5 = quantile(lowest_indel_count_driver$lowest_indel_count, 0.5),
                                  perc0.75 = quantile(lowest_indel_count_driver$lowest_indel_count, 0.75),
                                  perc0.95 = quantile(lowest_indel_count_driver$lowest_indel_count, 0.95))

#Bind rows
minimum_indel_summary_table <- bind_rows(distribution_per_sample,distribution_per_project,distribution_per_common_tumor,distribution_per_driver)


print(minimum_indel_summary_table)

```

# Tumor biases (per driver gene)
```{r}
mutant_total_indels <- mutant_indel_data %>% 
  filter(!TYPE %in% multiple_drivers_data$TYPE) %>%
  dplyr::group_by(driver,sample) %>%
  dplyr::summarise(sum_CILAD_NHEJ = sum(CILAD_NHEJ),
                   sum_CILAD_MMEJ = sum(CILAD_MMEJ),
                   sum_CLAD_NHEJ = sum(CLAD_NHEJ),
                   sum_CLAD_MMEJ = sum(CLAD_MMEJ),
                   sample_n = n()) %>%
  rowwise() %>%
  mutate(LAD_bal = log2(sum_CLAD_MMEJ/sum_CLAD_NHEJ),
                 iLAD_bal = log2(sum_CILAD_MMEJ/sum_CILAD_NHEJ),
                  global_bal = log2((sum_CLAD_MMEJ+sum_CILAD_MMEJ)/(sum_CLAD_NHEJ+sum_CILAD_NHEJ)),
                 bias = LAD_bal - iLAD_bal,
         total_indel = sum(sum_CILAD_NHEJ,sum_CILAD_MMEJ, sum_CLAD_NHEJ,sum_CLAD_MMEJ))

control_total_indels <- control_indel_data %>% 
  dplyr::group_by(driver,sample) %>%
  dplyr::summarise(sum_CILAD_NHEJ = sum(CILAD_NHEJ),
                   sum_CILAD_MMEJ = sum(CILAD_MMEJ),
                   sum_CLAD_NHEJ = sum(CLAD_NHEJ),
                   sum_CLAD_MMEJ = sum(CLAD_MMEJ),
                   sample_n = n()) %>%
  rowwise() %>%
  mutate(LAD_bal = log2(sum_CLAD_MMEJ/sum_CLAD_NHEJ),
                 iLAD_bal = log2(sum_CILAD_MMEJ/sum_CILAD_NHEJ),
                  global_bal = log2((sum_CLAD_MMEJ+sum_CILAD_MMEJ)/(sum_CLAD_NHEJ+sum_CILAD_NHEJ)),
                 bias = LAD_bal - iLAD_bal,
         total_indel = sum(sum_CILAD_NHEJ,sum_CILAD_MMEJ, sum_CLAD_NHEJ,sum_CLAD_MMEJ))

#bind_rows
total_changes_indels <- mutant_total_indels %>% bind_rows(control_total_indels)
  
```

# Tumor biases (per driver genecommon tumor type)
```{r}
mutant_total_indels_common <- mutant_indel_data %>% 
  filter(!TYPE %in% multiple_drivers_data$TYPE) %>%
  separate(project, into =c("tumor", "country")) %>%
  right_join(tumor_types_cases) %>%
  dplyr::group_by(driver,tumor,sample) %>%
  dplyr::summarise(sum_CILAD_NHEJ = sum(CILAD_NHEJ),
                   sum_CILAD_MMEJ = sum(CILAD_MMEJ),
                   sum_CLAD_NHEJ = sum(CLAD_NHEJ),
                   sum_CLAD_MMEJ = sum(CLAD_MMEJ),
                   sample_n = n()) %>%
  rowwise() %>%
  mutate(LAD_bal = log2(sum_CLAD_MMEJ/sum_CLAD_NHEJ),
                 iLAD_bal = log2(sum_CILAD_MMEJ/sum_CILAD_NHEJ),
                  global_bal = log2((sum_CLAD_MMEJ+sum_CILAD_MMEJ)/(sum_CLAD_NHEJ+sum_CILAD_NHEJ)),
                 bias = LAD_bal - iLAD_bal,
         total_indel = sum(sum_CILAD_NHEJ,sum_CILAD_MMEJ, sum_CLAD_NHEJ,sum_CLAD_MMEJ))

control_total_indels_common <- control_indel_data %>% 
    separate(project, into =c("tumor", "country")) %>%
  right_join(tumor_types_cases) %>%
  dplyr::group_by(driver,tumor,sample) %>%
  dplyr::summarise(sum_CILAD_NHEJ = sum(CILAD_NHEJ),
                   sum_CILAD_MMEJ = sum(CILAD_MMEJ),
                   sum_CLAD_NHEJ = sum(CLAD_NHEJ),
                   sum_CLAD_MMEJ = sum(CLAD_MMEJ),
                   sample_n = n()) %>%
  rowwise() %>%
  mutate(LAD_bal = log2(sum_CLAD_MMEJ/sum_CLAD_NHEJ),
                 iLAD_bal = log2(sum_CILAD_MMEJ/sum_CILAD_NHEJ),
                  global_bal = log2((sum_CLAD_MMEJ+sum_CILAD_MMEJ)/(sum_CLAD_NHEJ+sum_CILAD_NHEJ)),
                 bias = LAD_bal - iLAD_bal,
         total_indel = sum(sum_CILAD_NHEJ,sum_CILAD_MMEJ, sum_CLAD_NHEJ,sum_CLAD_MMEJ))

#bind_rows
total_changes_indels_common <- mutant_total_indels_common %>% bind_rows(control_total_indels_common)
  
```
#Data table averaged by tumor type with more than 5 patients.


#General trends: per driver
#Global changes
```{r}
ggplot(total_changes_indels) + 
  geom_col(aes(driver, global_bal, fill = sample), position = "dodge") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust = 0))

```

#Chromatin bias: LAD compared to iLAD (negative value mean bias)
```{r}
driver_order <- tibble(driver = c("ATM","BRCA2","MEN1","ATRX","TRRAP","BRCA1","SETD2","EGFR"), trend_validates = c("similar_trend","similar_trend","similar_trend","similar_trend","not_clear","not_clear","opposite_trend","opposite_trend"))

#table
plot_table <- total_changes_indels %>% left_join(driver_order)

ggplot(plot_table %>% filter(driver != "BRCA1")) + 
  geom_col(aes(fct_relevel(driver, driver_order$driver), bias, fill = sample), position = "dodge") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust = 0.5,vjust = 1),
        panel.grid = element_blank()) +
  geom_hline(yintercept = 0, linetype = 2) + ylab("Chromatin bias") +
  facet_wrap(~ fct_relevel(trend_validates, c("similar_trend","not_clear","opposite_trend")), scales = "free_x")
  

```

#Fisher test as we did it in the manuscript
```{r}
#Start with ATM cases
contingency_table <- total_changes_indels %>%
  filter(driver != "BRCA1") %>%
  select(driver, sample, sum_CILAD_MMEJ,sum_CILAD_NHEJ,sum_CLAD_MMEJ,sum_CLAD_NHEJ) %>%
  reshape2::melt() %>%
  separate(variable, into = c(NA, "chromatin","pathway")) %>%
  reshape2::dcast(driver + sample +pathway ~ chromatin)

#Unique combinations 
combinations <- contingency_table %>% select(driver, sample) %>% distinct()


#Function to assess significance

chq_test_summary <- map2_dfr(combinations$driver, combinations$sample, function(x,y){
  contingency_table %>% 
  filter(driver == x & sample == y) %>%
  select(-driver, -sample) %>%
  column_to_rownames(var = "pathway") %>% fisher_test() %>% mutate(driver = x, sample = y)
}) %>% mutate(fdr = p.adjust(p, method = "BH"))

```


# Null distribution (is it normally distributed?)
```{r,message=F,warning=F}
#Are they normal?
mutant_bias <- total_changes_indels %>% filter(sample == "mutant") %>% select(driver, sample_n,bias)

#bootstrap_general_tumors
boot_strap_total_rough <- map_dfr(mutant_bias$driver, function(x) {
  driver_dt <- mutant_bias %>% filter(driver == x)
  iteration_tib <- map_dfr(c(1:1000), function(j) {
    set.seed(j)
      control_indel_data %>% 
        filter(driver == x) %>%
        ungroup() %>%
        sample_n(driver_dt$sample_n) %>%
        mutate(iteration = j, driver = x) %>% 
          dplyr::group_by(iteration, driver) %>%
          dplyr::summarise(LAD_MMEJ = sum(CLAD_MMEJ), 
                           LAD_NHEJ = sum(CLAD_NHEJ), 
                           iLAD_MMEJ = sum(CILAD_MMEJ), 
                           iLAD_NHEJ = sum(CILAD_NHEJ), 
                           total_mutation = sum(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal,
                           diff_bias = driver_dt$bias - bias)
        })
      })

##Plot null distribution
ggplot(boot_strap_total_rough) + geom_density(aes(bias)) + facet_wrap(~ driver)

##Normality test
normality_test <- boot_strap_total_rough %>% dplyr::group_by(driver) %>% rstatix::shapiro_test(bias)

#diff analysis
ggplot(boot_strap_total_rough) + geom_density(aes(diff_bias)) +
  facet_wrap(~ driver) +
  geom_vline(xintercept = 0, linetype = 2) +
  theme_bw()

#Significance in the data
summary_permutation_test <- boot_strap_total_rough %>%
  ungroup() %>%
  dplyr::group_by(driver) %>%
  dplyr::summarise(permutations_under_0 = sum(diff_bias < 0)/1000)

#Calculate z-scores
z_score_calculation <- mutant_bias %>% mutate(z_score = (bias - mean(boot_strap_total_rough$bias))/sd(boot_strap_total_rough$bias))

```

# Null distribution (with common tumors)
```{r,message=F,warning=F}
#Are they normal?
mutant_bias_common_tumor <- total_changes_indels_common %>% filter(sample == "mutant") %>% select(driver, tumor, sample_n,bias)  %>% right_join(tumor_types_cases[c(1,2,4:11),])
control_indel_data_common_tumor <- control_indel_data %>% separate(project, into = c("tumor", "country")) %>% right_join(tumor_types_cases[c(1,2,4:11),])

#bootstrap_general_tumors
boot_strap_total_rough <- map2_dfr(mutant_bias_common_tumor$driver, mutant_bias_common_tumor$tumor, function(x,y) {
  driver_dt <- mutant_bias_common_tumor %>% filter(driver == x & tumor == y) 
  control_samples <- control_indel_data_common_tumor %>% filter(driver == x & tumor == y)
  
  iteration_tib <- map_dfr(c(1:1000), function(j) {
    set.seed(j)
      control_samples %>% 
        ungroup() %>%
        sample_n(driver_dt$sample_n) %>%
        mutate(iteration = j, driver = x, tumor = y) %>% 
          dplyr::group_by(iteration, driver, tumor) %>%
          dplyr::summarise(LAD_MMEJ = sum(CLAD_MMEJ), 
                           LAD_NHEJ = sum(CLAD_NHEJ), 
                           iLAD_MMEJ = sum(CILAD_MMEJ), 
                           iLAD_NHEJ = sum(CILAD_NHEJ), 
                           total_mutation = sum(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal,
                           diff_bias = driver_dt$bias - bias)
        })
      })

##Plot null distribution
ggplot(boot_strap_total_rough) + geom_density(aes(bias)) + facet_wrap(~ driver+tumor)

##Normality test
normality_test <- boot_strap_total_rough %>% dplyr::group_by(driver,tumor) %>% rstatix::shapiro_test(bias)

#diff analysis
ggplot(boot_strap_total_rough) + geom_density(aes(diff_bias)) +
  facet_wrap(~ driver+tumor) +
  geom_vline(xintercept = 0, linetype = 2) +
  theme_bw()

#Significance in the data
summary_permutation_test <- boot_strap_total_rough %>%
  ungroup() %>%
  dplyr::group_by(driver,tumor) %>%
  dplyr::summarise(permutations_under_0 = sum(diff_bias < 0)/100)

#Calculate z-scores
z_score_calculation <- mutant_bias %>% mutate(z_score = (bias - mean(boot_strap_total_rough$bias))/sd(boot_strap_total_rough$bias))

```


#Permutation test per driver (for ATM first)
```{r, message=F, warning=F}
#Are they normal?
ATM_data <- all_indel_data_together %>% filter(driver == "ATM")
case_control_files <- ATM_data %>% dplyr::group_by(sample) %>% dplyr::summarise(counts = n())

#Experimental difference
experimental_bias <- ATM_data %>% 
  dplyr::group_by(driver,sample) %>% 
  dplyr::summarise(LAD_MMEJ = sum(CLAD_MMEJ), 
                           LAD_NHEJ = sum(CLAD_NHEJ), 
                           iLAD_MMEJ = sum(CILAD_MMEJ), 
                           iLAD_NHEJ = sum(CILAD_NHEJ), 
                           total_mutation = sum(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal)

#Experimental bias (-0.446)

#bootstrap_general_tumors
boot_strap_total_rough <- map_dfr("ATM", function(x) {
  driver_dt <- ATM_data %>% filter(driver == x)
  cases_n <- driver_dt %>% filter(sample == "mutant") %>% nrow()
  iteration_tib <- map_dfr(c(1:2000), function(j) {
    set.seed(j)
      cases_samples <- driver_dt %>% 
        ungroup() %>%
        sample_n(cases_n) %>%
        pull(TYPE)
      
      driver_dt %>% mutate(perm = case_when(TYPE %in% cases_samples ~ "cases",
                                              T ~ "controls"),
                           iteration = j) %>%
        dplyr::group_by(driver,perm, iteration) %>% 
        dplyr::summarise(LAD_MMEJ = sum(CLAD_MMEJ), 
                           LAD_NHEJ = sum(CLAD_NHEJ), 
                           iLAD_MMEJ = sum(CILAD_MMEJ), 
                           iLAD_NHEJ = sum(CILAD_NHEJ), 
                           total_mutation = sum(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal)
        })
      })

#Calculate difference
difference_summary <- boot_strap_total_rough %>% reshape2::dcast(driver+iteration ~ perm, value.var = "bias") %>% mutate(perm_diff = cases - controls)

##Plot null distribution
ggplot(boot_strap_total_rough) + geom_density(aes(bias)) + facet_wrap(~ driver) 

##Normality test
normality_test <- boot_strap_total_rough %>% dplyr::group_by(driver) %>% rstatix::shapiro_test(bias)

#diff analysis
ggplot(boot_strap_total_rough) + geom_density(aes(diff_bias)) +
  facet_wrap(~ driver) +
  geom_vline(xintercept = 0, linetype = 2) +
  theme_bw()

#Significance in the data
summary_permutation_test <- boot_strap_total_rough %>%
  ungroup() %>%
  dplyr::group_by(driver) %>%
  dplyr::summarise(permutations_under_0 = sum(diff_bias < 0)/1000)

#Calculate z-scores
z_score_calculation <- mutant_bias %>% mutate(z_score = (bias - mean(boot_strap_total_rough$bias))/sd(boot_strap_total_rough$bias))





```



# Null distribution (is it normally distributed?)
```{r,message=F,warning=F}
#Are they normal?
mutant_bias <- total_changes_indels %>% filter(sample == "mutant") %>% select(driver, sample_n,bias)

#bootstrap_general_tumors
boot_strap_total_rough <- map_dfr(mutant_bias$driver, function(x) {
  driver_dt <- mutant_bias %>% filter(driver == x)
  iteration_tib <- map_dfr(c(1:1000), function(j) {
    set.seed(j)
      control_indel_data %>% 
        
        ungroup() %>%
        sample_n(driver_dt$sample_n) %>%
        mutate(iteration = j, driver = x) %>% 
          dplyr::group_by(iteration, driver) %>%
          dplyr::summarise(LAD_MMEJ = sum(CLAD_MMEJ), 
                           LAD_NHEJ = sum(CLAD_NHEJ), 
                           iLAD_MMEJ = sum(CILAD_MMEJ), 
                           iLAD_NHEJ = sum(CILAD_NHEJ), 
                           total_mutation = sum(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal,
                           diff_bias = driver_dt$bias - bias)
        })
      })

##Plot null distribution
ggplot(boot_strap_total_rough) + geom_density(aes(bias)) + facet_wrap(~ driver)

##Normality test
normality_test <- boot_strap_total_rough %>% dplyr::group_by(driver) %>% rstatix::shapiro_test(bias)

#diff analysis
ggplot(boot_strap_total_rough) + geom_density(aes(diff_bias)) +
  facet_wrap(~ driver) +
  geom_vline(xintercept = 0, linetype = 2) +
  theme_bw()

#Significance in the data
boot_strap_total_rough %>%
  dplyr::group_by(driver) %>%
  dplyr::summarise(CI95 = quantile(diff_bias, 0.9))

#Calculate z-scores
z_score_calculation <- mutant_bias %>% mutate(z_score = (bias - mean(boot_strap_total_rough$bias))/sd(boot_strap_total_rough$bias))

```

# Null distribution (is it normally distributed?)
```{r,message=F,warning=F}
#Are they normal?
mutant_bias <- total_changes_indels %>% filter(sample == "mutant") %>% select(driver, sample_n,bias)

#bootstrap_general_tumors
boot_strap_total_rough <- map_dfr(mutant_bias$driver, function(x) {
  driver_dt <- mutant_bias %>% filter(driver == x)
  iteration_tib <- map_dfr(c(1:1000), function(j) {
    set.seed(j)
      control_indel_data %>% 
        ungroup() %>%
        sample_n(driver_dt$sample_n) %>%
        mutate(iteration = j, driver = x) %>% 
          dplyr::group_by(iteration, driver) %>%
          dplyr::summarise(LAD_MMEJ = sum(CLAD_MMEJ), 
                           LAD_NHEJ = sum(CLAD_NHEJ), 
                           iLAD_MMEJ = sum(CILAD_MMEJ), 
                           iLAD_NHEJ = sum(CILAD_NHEJ), 
                           total_mutation = sum(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal,
                           diff_bias = driver_dt$bias - bias)
        })
      })

##Plot null distribution
ggplot(boot_strap_total_rough) + geom_density(aes(bias)) + facet_wrap(~ driver)

##Normality test
normality_test <- boot_strap_total_rough %>% dplyr::group_by(driver) %>% rstatix::shapiro_test(bias)

#diff analysis
ggplot(boot_strap_total_rough) + geom_density(aes(diff_bias)) +
  facet_wrap(~ driver) +
  geom_vline(xintercept = 0, linetype = 2) +
  theme_bw()

#Significance in the data
boot_strap_total_rough %>%
  dplyr::group_by(driver) %>%
  dplyr::summarise(CI95 = quantile(diff_bias, 0.95))

#Calculate z-scores
z_score_calculation <- mutant_bias %>% mutate(z_score = (bias - mean(boot_strap_total_rough$bias))/sd(boot_strap_total_rough$bias))

```


# Tumor biases (per project)
```{r}
mutant_total_indels_project <- mutant_indel_data %>% 
    filter(!TYPE %in% multiple_drivers_data$TYPE) %>%
  dplyr::group_by(project,driver,sample) %>%
  dplyr::summarise(sum_CILAD_NHEJ = sum(CILAD_NHEJ),
                   sum_CILAD_MMEJ = sum(CILAD_MMEJ),
                   sum_CLAD_NHEJ = sum(CLAD_NHEJ),
                   sum_CLAD_MMEJ = sum(CLAD_MMEJ),
                   sample_n = n()) %>%
  rowwise() %>%
  mutate(LAD_bal = log2(sum_CLAD_MMEJ/sum_CLAD_NHEJ),
                 iLAD_bal = log2(sum_CILAD_MMEJ/sum_CILAD_NHEJ),
         global_bal = log2((sum_CLAD_MMEJ+sum_CILAD_MMEJ)/(sum_CLAD_NHEJ+sum_CILAD_NHEJ)),
                 bias = LAD_bal - iLAD_bal,
         total_indel = sum(sum_CILAD_NHEJ,sum_CILAD_MMEJ, sum_CLAD_NHEJ,sum_CLAD_MMEJ))

control_total_indels_project <- control_indel_data %>% 
  dplyr::group_by(project,driver,sample) %>%
  dplyr::summarise(sum_CILAD_NHEJ = sum(CILAD_NHEJ),
                   sum_CILAD_MMEJ = sum(CILAD_MMEJ),
                   sum_CLAD_NHEJ = sum(CLAD_NHEJ),
                   sum_CLAD_MMEJ = sum(CLAD_MMEJ),
                   sample_n = n()) %>%
  rowwise() %>%
  mutate(LAD_bal = log2(sum_CLAD_MMEJ/sum_CLAD_NHEJ),
                 iLAD_bal = log2(sum_CILAD_MMEJ/sum_CILAD_NHEJ),
         global_bal = log2((sum_CLAD_MMEJ+sum_CILAD_MMEJ)/(sum_CLAD_NHEJ+sum_CILAD_NHEJ)),
                 bias = LAD_bal - iLAD_bal,
         total_indel = sum(sum_CILAD_NHEJ,sum_CILAD_MMEJ, sum_CLAD_NHEJ,sum_CLAD_MMEJ))

#bind_rows
total_changes_indels_project <- mutant_total_indels_project %>% bind_rows(control_total_indels_project)
  
```

# Tumor biases (per tumor type)
```{r}
mutant_total_indels_project <- mutant_indel_data %>% 
  separate(project, into = c("tumor", "country"), remove = F) %>%
    filter(!TYPE %in% multiple_drivers_data$TYPE) %>%
  dplyr::group_by(tumor,driver,sample) %>%
  dplyr::summarise(sum_CILAD_NHEJ = sum(CILAD_NHEJ),
                   sum_CILAD_MMEJ = sum(CILAD_MMEJ),
                   sum_CLAD_NHEJ = sum(CLAD_NHEJ),
                   sum_CLAD_MMEJ = sum(CLAD_MMEJ),
                   sample_n = n()) %>%
  rowwise() %>%
  mutate(LAD_bal = log2(sum_CLAD_MMEJ/sum_CLAD_NHEJ),
                 iLAD_bal = log2(sum_CILAD_MMEJ/sum_CILAD_NHEJ),
         global_bal = log2((sum_CLAD_MMEJ+sum_CILAD_MMEJ)/(sum_CLAD_NHEJ+sum_CILAD_NHEJ)),
                 bias = LAD_bal - iLAD_bal,
         total_indel = sum(sum_CILAD_NHEJ,sum_CILAD_MMEJ, sum_CLAD_NHEJ,sum_CLAD_MMEJ)) %>% filter(sample_n > 5 & total_indel > 50)

#Filter controls for these tumors
tumor_types_cases <- mutant_total_indels_project %>% select(driver, tumor) %>% unique()


control_total_indels_project <- control_indel_data %>% 
  separate(project, into = c("tumor", "country"), remove = F) %>%
  dplyr::group_by(tumor,driver,sample) %>%
  dplyr::summarise(sum_CILAD_NHEJ = sum(CILAD_NHEJ),
                   sum_CILAD_MMEJ = sum(CILAD_MMEJ),
                   sum_CLAD_NHEJ = sum(CLAD_NHEJ),
                   sum_CLAD_MMEJ = sum(CLAD_MMEJ),
                   sample_n = n()) %>%
  rowwise() %>%
  mutate(LAD_bal = log2(sum_CLAD_MMEJ/sum_CLAD_NHEJ),
                 iLAD_bal = log2(sum_CILAD_MMEJ/sum_CILAD_NHEJ),
         global_bal = log2((sum_CLAD_MMEJ+sum_CILAD_MMEJ)/(sum_CLAD_NHEJ+sum_CILAD_NHEJ)),
                 bias = LAD_bal - iLAD_bal,
         total_indel = sum(sum_CILAD_NHEJ,sum_CILAD_MMEJ, sum_CLAD_NHEJ,sum_CLAD_MMEJ)) %>%
  right_join(tumor_types_cases)

#bind_rows
total_changes_indels_project <- mutant_total_indels_project %>% bind_rows(control_total_indels_project)
  
```

#Control distributions of mutation
```{r}
ggplot(total_changes_indels_sample) + 
  geom_density(aes(log10(total_indel), color = sample), position = "dodge") +
  facet_wrap(~ driver) + theme_bw()

#How many total mutations?
ggplot(total_changes_indels_sample) + 
  geom_quasirandom(aes(sample, min_indel, color = sample)) +
  facet_wrap(~ driver+tumor, scales = "free") + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

ggplot(total_changes_indels_sample) + 
  geom_quasirandom(aes(sample, bias, color = sample)) +
  facet_wrap(~ driver+tumor, scales = "free") + theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

```


#Plot global changes per project
```{r}
ggplot(total_changes_indels_project) + 
  geom_col(aes(tumor, global_bal, fill = sample), position = "dodge") + 
  facet_wrap(~ driver, scales = "free") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust = 0))

```

#Plot global changes per project
```{r}
ggplot(total_changes_indels_project) + 
  geom_col(aes(tumor, bias, fill = sample), position = "dodge") + 
  facet_wrap(~ driver, scales = "free") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust = 0)) +
  geom_hline(yintercept = 0, linetype = 2)

```

#Per tumor type: only if enough data
```{r, message = F, warning = F}
#Set up randomization test
#Mutant samples, tumor distribution
tumor_distribution <- mutant_indel_data %>% 
  separate(project, into = c("tumor","country"), remove = F) %>%
  right_join(tumor_types_cases) %>%
  dplyr::group_by(driver,tumor) %>% dplyr::summarise(tumor_c = n())

control_distribution <- control_indel_data %>% 
  separate(project, into = c("tumor","country"), remove = F) %>%
  right_join(tumor_types_cases) %>%
  dplyr::group_by(driver,tumor) %>% dplyr::summarise(control_c = n())

tumor_distribution <- left_join(tumor_distribution, control_distribution) %>% filter(tumor_c < control_c)


boot_strap_total_indels <- map2_dfr(tumor_distribution$driver,tumor_distribution$tumor, function(x,y) {
  driver_dt <- tumor_distribution %>% filter(driver == x & tumor == y) %>% pull(tumor_c)
  controls_dt <- control_total_indels_sample %>% separate(project, into = c("tumor","country"), remove = F) %>%
  right_join(tumor_types_cases) %>% filter(driver == x & tumor == y) 
  iteration_tib <- map_dfr(c(1:1000), function(j) {
    set.seed(j)
      control_samples <- controls_dt %>%
          sample_n(driver_dt) %>% 
          mutate(iteration = j, driver = x, tumor = y) %>% 
          dplyr::group_by(iteration, driver, tumor) %>%
          dplyr::summarise(LAD_MMEJ = median(CLAD_MMEJ), 
                           LAD_NHEJ = median(CLAD_NHEJ), 
                           iLAD_MMEJ = median(CILAD_MMEJ), 
                           iLAD_NHEJ = median(CILAD_NHEJ), 
                           total_mutation = median(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal)
        })
      })

#Null distribution summary
null_distribution <- boot_strap_total_indels %>% dplyr::group_by(tumor,driver) %>% dplyr::summarise(mean = mean(bias), sd = sd(bias))

#Left join
total_numbers <- total_changes_indels_project %>% left_join(null_distribution) %>% mutate(z_score = (bias - mean)/sd) %>% na.omit()

```

#Rough permutation test with all the data
```{r, warning=F, message=F}
mutant_bias <- total_changes_indels %>% filter(sample == "mutant") %>% select(driver, sample_n,bias)

#bootstrap_general_tumors
boot_strap_total_rough <- map_dfr(mutant_bias$driver, function(x) {
  driver_dt <- mutant_bias %>% filter(driver == x)
  iteration_tib <- map_dfr(c(1:1000), function(j) {
    set.seed(j)
      control_total_indels_sample %>% 
        ungroup() %>%
        sample_n(driver_dt$sample_n) %>%
        mutate(iteration = j, driver = x) %>% 
          dplyr::group_by(iteration, driver) %>%
          dplyr::summarise(LAD_MMEJ = sum(CLAD_MMEJ), 
                           LAD_NHEJ = sum(CLAD_NHEJ), 
                           iLAD_MMEJ = sum(CILAD_MMEJ), 
                           iLAD_NHEJ = sum(CILAD_NHEJ), 
                           total_mutation = sum(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal,
                           diff_bias = driver_dt$bias - bias)
        })
      })

#ggplot difference densities
ggplot(boot_strap_total_rough) + 
  geom_density(aes(diff_bias)) + 
  facet_wrap(~ driver) + 
  geom_vline(xintercept = 0, linetype =2) +
  theme_bw()

#Summarize the 0.95 value
permutation_test_result <- boot_strap_total_rough %>% 
  dplyr::group_by(driver) %>% 
  dplyr::summarise(CI95 = quantile(diff_bias, 0.95),
                   CI90 = quantile(diff_bias, 0.9))

```


#Filter projects 
```{r}
plot_filter_ATM <- total_changes_indels_project %>% filter(driver == "ATM" & !project %in% c("ESAD-UK","PAEN-AU","CLLE-ES"))

ggplot(plot_filter_ATM) + geom_point(aes(sample,bias)) + geom_line(aes(sample,bias, group = project))

tmp <- plot_filter_ATM %>% select(project,bias, sample) %>% reshape2::dcast(project~sample, value.var = "bias")


t.test(tmp$control, tmp$mutant, paired = T)

```

```{r}
plot_filter_ATRX <- total_changes_indels_project %>% filter(driver == "ATRX")

ggplot(plot_filter_ATRX) + geom_point(aes(sample,bias)) + geom_line(aes(sample,bias, group = project))

tmp <- plot_filter_ATRX %>% select(project,bias, sample) %>% reshape2::dcast(project~sample, value.var = "bias")


t.test(tmp$control, tmp$mutant, paired = T)
```



#Plot global changes
```{r}
ggplot(total_changes_indels) + geom_col(aes(sample,bias)) + 
  facet_wrap(~ driver) + 
  geom_hline(yintercept = 0, linetype = 2) +
  theme_bw() +
  theme(panel.grid = element_blank())


#Calculate delta
delta_calculate <- total_changes_indels %>% 
  reshape2::dcast(driver ~ sample, value.var = "global_bal") %>%
  mutate(delta_global = mutant - control)

ggplot(total_changes_indels) + geom_col(aes(sample, global_bal)) + 
  facet_wrap(~ driver) + 
  geom_hline(yintercept = 0, linetype = 2) +
  theme_bw() +
  theme(panel.grid = element_blank()) + ylab("∆log2MMEJ:NHEJ")
```
#Conclusion: Several of these genes show similar trends. For example: ATM, ATRX (this one is very cool!), BRCA2 and MEN1.
#Others don't validate. Again chromatin associated proteins or proteins that are likely indirect (SETD2 and EGFR)


#Re-do plots as in the manuscript, per protein
```{r, fig.width=3, fig.height=5}
manuscript_plot <- total_changes_indels %>% select(driver,sample,LAD_bal,iLAD_bal) %>% melt()

#Similar trend (ATM, BRCA2, MEN1 and ATRX)
ggplot(manuscript_plot %>% filter(driver %in% c("ATM","BRCA2","ATRX","MEN1"))) + 
  geom_col(aes(fct_relevel(variable,c("iLAD_bal","LAD_bal")),value)) + 
  facet_grid(driver ~ sample) +
  theme_bw() + ylab("log2MMEJ:NHEJ") +
  theme(panel.grid = element_blank(), axis.title.x = element_blank())

#Different trend (SETD2 and EGFR)
ggplot(manuscript_plot %>% filter(driver %in% c("SETD2","EGFR"))) + 
  geom_col(aes(fct_relevel(variable,c("iLAD_bal","LAD_bal")),value)) + 
  facet_grid(driver ~ sample) +
  theme_bw() + ylab("log2MMEJ:NHEJ") +
  theme(panel.grid = element_blank(), axis.title.x = element_blank())

#Difficult interpretation (BRCA1 and TRRAP)
ggplot(manuscript_plot %>% filter(driver %in% c("BRCA1","TRRAP"))) + 
  geom_col(aes(fct_relevel(variable,c("iLAD_bal","LAD_bal")),value)) + 
  facet_grid(driver ~ sample) +
  theme_bw() + ylab("log2MMEJ:NHEJ") +
  theme(panel.grid = element_blank(), axis.title.x = element_blank())


```

#Calculate deltas as I calculate in the cells
```{r}
# Interpretable plot
delta_log2 <- total_changes_indels %>% 
  select(driver, sample, LAD_bal,iLAD_bal) %>% 
  reshape2::melt() %>% 
  reshape2::dcast(driver ~ sample+variable) %>%
  mutate(delta_LAD = mutant_LAD_bal - control_LAD_bal,
         delta_iLAD = mutant_iLAD_bal - control_iLAD_bal)

# Interpretable plot
delta_log2_projects <- total_changes_indels_project %>% 
  select(driver, sample,project, LAD_bal,iLAD_bal) %>% 
  reshape2::melt() %>% 
  reshape2::dcast(driver+project ~ sample+variable) %>%
  mutate(delta_LAD = mutant_LAD_bal - control_LAD_bal,
         delta_iLAD = mutant_iLAD_bal - control_iLAD_bal)

table_plot_delta_chromatin <- delta_log2 %>%
  select(driver, delta_LAD, delta_iLAD) %>%
  melt()

#
ggplot(table_plot_delta_chromatin) +
  geom_col(aes(fct_relevel(variable, c("delta_iLAD","delta_cLAD")), value)) +
  facet_wrap(~ driver, ncol = 4) +
  ylab("∆log2 MMEJ:NHEJ") +
  theme_bw() +
  geom_hline(yintercept = 0, linetype = 2)
```


#Set up randomization test
```{r}
#Mutant samples, tumor distribution
tumor_distribution <- mutant_indel_data %>% dplyr::group_by(driver,project) %>% dplyr::summarise(tumor_c = n()) %>% filter(!project %in% c("ESAD-UK","PAEN-AU","CLLE-ES"))
control_distribution <- control_indel_data %>% dplyr::group_by(driver,project) %>% dplyr::summarise(control_c = n())

boot_strap_total_indels <- map_dfr("ATM", function(x) {
  driver_dt <- tumor_distribution %>% filter(driver == x)
  controls_dt <- control_indel_data %>% filter(driver == x)
  iteration_tib <- map_dfr(c(1:100), function(j) {
    set.seed(j)
      control_samples <- map2_dfr(driver_dt$project,driver_dt$tumor_c, function(y,z) {
      controls_dt %>% 
          filter(project == y) %>% 
          sample_n(z) %>% 
          mutate(iteration = j, driver = x) %>% 
          dplyr::summarise(LAD_MMEJ = sum(CLAD_MMEJ), 
                           LAD_NHEJ = sum(CLAD_NHEJ), 
                           iLAD_MMEJ = sum(CILAD_MMEJ), 
                           iLAD_NHEJ = sum(CILAD_NHEJ), 
                           total_mutation = sum(CILAD_NHEJ +CILAD_MMEJ+CLAD_NHEJ+CLAD_MMEJ), 
                           sample_n = n(),
                           LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
                           iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
                           bias = LAD_bal - iLAD_bal,
                           project = y)
        })
  #    tibble(iteration = j, median_bal = median(control_samples$bias))
      })
  })

#Compute metric per iteration
summary_per_iteration <- boot_strap_total_indels %>%
  mutate(LAD_bal = log2(LAD_MMEJ/LAD_NHEJ),
         iLAD_bal = log2(iLAD_MMEJ/iLAD_NHEJ),
         bias = LAD_bal - iLAD_bal)

#Mean and sd of these mutations
summary_null_filter <- summary_per_iteration %>% 
  filter(total_mutation > 17) %>%
  dplyr::group_by(driver) %>%
  dplyr::summarise(mean = mean(bias),
                   sd = sd(bias),
                   sample_n = n())

summary_null <- summary_per_iteration %>% 
  dplyr::group_by(driver) %>%
  dplyr::summarise(mean = mean(bias),
                   sd = sd(bias),
                   sample_n = n())

```

#Summary numbers cases
```{r}
#Summary table with n
summary_sample_n_cases <- mutant_indel_data %>%
  dplyr::group_by(driver,project) %>%
  dplyr::summarise(sample_n = n()/2)

#Filtering strategy 
filtered_samples_cases <- mutant_indel_data %>%
  filter(NHEJ + MMEJ > 2) %>%
  dplyr::group_by(TYPE, driver, project) %>%
  dplyr::summarise(count = n()) %>%
  filter(count == 2) %>%
  ungroup() %>%
    dplyr::group_by(driver,project) %>%
  dplyr::summarise(kept_n = n())

# Filter
ggplot(summary_sample_n_cases) +
  geom_col(aes(project, sample_n, fill = driver), position = "dodge") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))

# Filter out percentage
ggplot(filtered_samples_cases) +
  geom_col(aes(project, kept_n, fill = driver), position = "dodge") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))

#mean control sample are 40
#median control samples are 33

```


#What is the percetage of samples discarded
```{r}
discarded_percentage_controls <- summary_sample_n %>% 
  left_join(filtered_samples) %>% 
  mutate(kept_percentage_controls = 100*(kept_n/sample_n))

discarded_percentage_cases <- summary_sample_n_cases %>% 
  left_join(filtered_samples_cases) %>% 
  mutate(kept_percentage_cases = 100*(kept_n/sample_n))

#correlation between them
plot_kept_percentage <-discarded_percentage_controls %>% 
  select(driver,project,kept_percentage_controls) %>% 
  left_join(discarded_percentage_cases%>% 
              select(driver,project,kept_percentage_cases))

#Plot percentage
ggplot(plot_kept_percentage,aes(kept_percentage_controls, kept_percentage_cases)) +
  geom_point() +
  theme_bw() +
  stat_cor() +
  geom_abline() +
  geom_smooth(method = "lm")
  

```

#log2 balance in the control samples (per sample)
```{r, warning=T, message=F}
log2_balance_data_sample <- map_dfr(c(0, 1,2,5,10,20,30,40,50), function(x) {
  control_indel_data %>%
  filter(NHEJ + MMEJ >= x) %>%
  dplyr::group_by(TYPE, driver, project) %>%
  dplyr::summarise(NHEJ = sum(NHEJ),
                   MMEJ = sum(MMEJ)) %>%
  mutate(log2_bal = log2((MMEJ+1)/(NHEJ+1)),
         filter_n = x)
})

#Mean_log2
mean_log2_bal <- log2_balance_data_sample %>%
  dplyr::group_by(driver,project, filter_n) %>%
  dplyr::summarise(mean_log2_bal = mean(log2_bal),
                   sample_n = n(),
                   median_NHEJ = median(NHEJ),
                   median_MMEJ = median(MMEJ))

 ggplot(log2_balance_data_sample,aes(project,log2_bal, color = log2(NHEJ+MMEJ))) + 
   stat_summary(geom = "col") +
   geom_point() +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 90)) + facet_wrap(~ filter_n)
```
#Do I see changes with my mutants (global changes)

#log2 balance in the control samples (per project)
```{r, warning=T, message=F}
log2_balance_data_project <- control_indel_data %>%
  dplyr::group_by(driver, project) %>%
  dplyr::summarise(NHEJ = sum(NHEJ),
                   MMEJ = sum(MMEJ)) %>%
  mutate(mean_log2_bal = log2((MMEJ+1)/(NHEJ+1)))

 ggplot(log2_balance_data_project,aes(project,mean_log2_bal, color = log2(NHEJ+MMEJ))) + 
   stat_summary(geom = "col") +
   geom_point() +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 90))
```

#Plot correlation of those samples
```{r}
summary_mean_controls <- mean_log2_bal %>%
  left_join(log2_balance_data_project %>% select(driver, project, project_sum_mean = mean_log2_bal))

ggplot(summary_mean_controls, aes(mean_log2_bal, project_sum_mean)) + 
  geom_point(aes(size = median_MMEJ)) + 
  stat_cor(method = "spearman") + 
  theme_bw() +
  geom_smooth(method = "lm") +
  xlab("log2_bal of total mutations") +
  ylab("Mean log2_bal of all samples") +
  facet_wrap(~ filter_n)
```
#Conclusion filtering for higher number doesn't help, high amount of mutations actually differs from sum

#Similar approach for chromatin bias
```{r}
chrom_bias_data_sample <- map_dfr(c(0:100), function(x) {
  control_indel_data %>%
  filter(NHEJ + MMEJ >= x) %>%
  dplyr::group_by(TYPE, driver, project,chromatin) %>%
  dplyr::summarise(log2_bal = log2((MMEJ+1)/(NHEJ + 1))) %>%
  reshape2::dcast(TYPE+driver+project ~ chromatin, value.var = "log2_bal") %>%
  mutate(filter_n = x)
}) %>% mutate(bias = CLAD - CILAD)

#Mean_log2
mean_bias <- chrom_bias_data_sample %>%
  dplyr::group_by(driver,project, filter_n) %>%
  dplyr::summarise(mean_bias = mean(bias, na.rm = T),
                   sample_n = n())

 ggplot(chrom_bias_data_sample,aes(project,bias)) + 
   stat_summary(geom = "col") +
#   geom_point() +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 90)) + facet_wrap(~ filter_n)


```

### Mean per tumor project
```{r}
control_data_mean_bias <- control_indel_data %>%
  dplyr::group_by(driver, project,chromatin) %>%
  dplyr::summarise(log2_bal = log2((sum(MMEJ)+1)/(sum(NHEJ) + 1))) %>%
  reshape2::dcast(driver+project ~ chromatin, value.var = "log2_bal")%>% mutate(bias = CLAD - CILAD)

 ggplot(control_data_mean_bias,aes(project,bias)) + 
   stat_summary(geom = "col") +
   geom_point() +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 90))

```

#Correlation between mean and filtered data
```{r}
summary_mean_controls_bias <- mean_bias %>%
  left_join(control_data_mean_bias %>% select(driver, project, project_sum_mean = bias))

#Cortest
summary_cortest <- summary_mean_controls_bias %>%
  dplyr::group_by(filter_n) %>%
  cor_test(mean_bias,project_sum_mean, method = "spearman")

#
ggplot(summary_cortest) + geom_line(aes(filter_n, -log10(p))) +
  theme_bw()

summary_cortest %>% slice_min(p) #10 indels total is the best correlation coefficient

```


```{r}
ggplot(summary_mean_controls_bias %>% filter(filter_n %in% c(0,1,5,10,20,25,50)), aes(mean_bias, project_sum_mean)) + 
  geom_point(aes(size = sample_n)) + 
  stat_cor(method = "spearman") + 
  theme_bw() +
  geom_smooth(method = "lm") +
  xlab("log2_bal of total mutations") +
  ylab("Mean log2_bal of all samples") +
  facet_wrap(~ filter_n)

```
#Conclusion: When it comes to the bias, the mean doesn't resemble to the mean of the tissue

#Measure mean values in mutated samples
```{r}
cases_data_mean_bias <- mutant_indel_data %>%
  dplyr::group_by(driver, project,chromatin) %>%
  dplyr::summarise(log2_bal = log2((sum(MMEJ)+1)/(sum(NHEJ) + 1))) %>%
  reshape2::dcast(driver+project ~ chromatin, value.var = "log2_bal")%>% mutate(bias = CLAD - CILAD)

#Match with controls
control_cases_summary <- cases_data_mean_bias %>%
  select(driver,project,case_bias = bias) %>%
  left_join(control_data_mean_bias %>%
              select(driver, project,control_bias = bias))

```

#Fisher test per tumor class
```{r}
#samples
BRCA2_cases_project <- mutant_indel_data %>% filter(driver == "BRCA2") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver+project~chromatin+variable, fun.aggregate = sum)  %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))
BRCA2_controls_project <- control_indel_data %>% filter(driver == "BRCA2") %>% mutate(driver = "control") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver+project~chromatin+variable, fun.aggregate = sum) %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))

#Plot
BRCA2_total_numbers <- bind_rows(BRCA2_cases_project, BRCA2_controls_project)

ggplot(BRCA2_total_numbers) + geom_quasirandom(aes(fct_relevel(driver, c("control","BRCA2")),bias)) + theme_bw()


wilcox_test <- BRCA2_total_numbers %>%
  wilcox_test(bias ~ driver,paired = T)

```


#BRCA2 total number of mutations (without different projects)
```{r}
#samples
BRCA2_cases_project <- mutant_indel_data %>% filter(driver == "BRCA2") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum)  %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))
BRCA2_controls_project <- control_indel_data %>% filter(driver == "BRCA2") %>% mutate(driver = "control") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum) %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))

#Plot
BRCA2_total_numbers <- bind_rows(BRCA2_cases_project, BRCA2_controls_project)

ggplot(BRCA2_total_numbers) + geom_col(aes(fct_relevel(driver, c("control","BRCA2")),bias)) + theme_bw()

```
#ATM total number of mutations (without different projects)
```{r}
#samples
ATM_cases_project <- mutant_indel_data %>% filter(driver == "ATM") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum)  %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))
ATM_controls_project <- control_indel_data %>% filter(driver == "ATM") %>% mutate(driver = "control") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum) %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))

#Plot
ATM_total_numbers <- bind_rows(ATM_cases_project, ATM_controls_project)

ggplot(ATM_total_numbers) + geom_col(aes(fct_relevel(driver, c("control","ATM")),bias)) + theme_bw() + geom_hline(yintercept = 0, linetype = 2)

```

#EGFR total number of mutations (without different projects)
```{r}
#samples
EGFR_cases_project <- mutant_indel_data %>% filter(driver == "EGFR") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum)  %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))
EGFR_controls_project <- control_indel_data %>% filter(driver == "EGFR") %>% mutate(driver = "control") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum) %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))

#Plot
EGFR_total_numbers <- bind_rows(EGFR_cases_project, EGFR_controls_project)

ggplot(EGFR_total_numbers) + geom_col(aes(fct_relevel(driver, c("control","ATM")),bias)) + theme_bw() + geom_hline(yintercept = 0, linetype = 2)

```


# Median indel amount per project in controls
```{r}
median_counts <- control_indel_data %>% 
  dplyr::group_by(driver, project, chromatin) %>%
  dplyr::summarise(median_prc = median(NHEJ+MMEJ),
                   n = n()/2)

filter_out_projects <- median_counts %>%
  filter(median_prc < 4) %>% select(driver, project) %>% distinct()

#Mean effect sizes
mean_effects_trends <- control_cases_summary %>% 
  anti_join(filter_out_projects) %>% 
  mutate(delta_effect = case_bias - control_bias)

#Mean effects per gene
mean_driver_effect <-  mean_effects_trends %>%
  dplyr::group_by(driver) %>%
  dplyr::summarise(mean_delta_driver = mean(delta_effect))

```

# Test boot-strapping in controls and build null distributions
```{r}
# Test for BRCA2 first
BRCA2_cases <- mutant_indel_data %>% filter(driver == "BRCA2") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(TYPE+driver+project~chromatin+variable)
BRCA2_controls <- control_indel_data %>% filter(driver == "BRCA2") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(TYPE+driver+project~chromatin+variable)

#Number of cases per tumor type 
n_cases_BRCA2 <- BRCA2_cases %>%
  select(TYPE,driver,project) %>%
  distinct() %>%
  dplyr::group_by(driver, project) %>% 
  dplyr::summarise(n = n())

#Number of controls per each tumor type (as QC)
n_controls_BRCA2 <- BRCA2_controls %>%
  select(TYPE,driver,project) %>%
  distinct() %>%
  dplyr::group_by(driver, project) %>% 
  dplyr::summarise(n = n())

#Bootstrap approach (before it repeats)
boot_strap_approach <- map2_dfr(n_cases_BRCA2$driver,n_cases_BRCA2$project, function(x,y){
  project_file <- BRCA2_controls %>% filter(driver == x & project == y)
  n_drawing <- n_cases_BRCA2 %>% filter(driver == x & project == y) %>% pull(n) %>% unique()
  map_dfr(c(1:2000), function(z){
    set.seed(z)
    project_file %>% 
      sample_n(n_drawing) %>%
      mutate(iteration = z) %>%
      dplyr::group_by(iteration) %>%
      dplyr::summarise(bal_LAD = log2((sum(CLAD_MMEJ))/(sum(CLAD_NHEJ))),
                       bal_iLAD = log2((sum(CILAD_MMEJ))/(sum(CILAD_NHEJ))),
                       bias = bal_LAD - bal_iLAD) %>%
      mutate(driver = x, project = y, sampling_n = n_drawing)
  })
})

#Calculate mean and sd of each approach
tmp <- boot_strap_approach %>%
  filter(is.finite(bias)) %>% dplyr::group_by(driver,project, sampling_n) %>%
  dplyr::summarise(mean_null = mean(bias, na.rm =T),
                  sd_null = sd(bias, na.rm = T)) %>% left_join(mean_effects_trends) %>%
  mutate(z_score = (case_bias - mean_null)/sd_null)

#Compute metric before
BRCA2_controls_metric <- BRCA2_controls %>% 
  rowwise() %>% 
  mutate(bal_LAD = log2((CLAD_MMEJ+1)/(CLAD_NHEJ+1)),
         bal_iLAD = log2((CILAD_MMEJ+1)/(CILAD_NHEJ+1)),
         bias = bal_LAD - bal_iLAD)

boot_strap_approach_mean <- map2_dfr(n_cases_BRCA2$driver,n_cases_BRCA2$project, function(x,y){
  project_file <- BRCA2_controls_metric %>% filter(driver == x & project == y)
#  n_drawing <- n_cases_BRCA2 %>% filter(driver == x & project == y) %>% pull(n) %>% unique()
#  map_dfr(c(1:100), function(z){
#    set.seed(z)
#    mean_bias <- project_file %>% 
#      sample_n(n_drawing) %>%
#      pull(bias) %>%
#      mean()
#    
#    tibble(driver = x, project = y, avg_bias = mean_bias,iteraction = z, n_sampling = n_drawing)
#  })
})



```

#Compute mean and sd per project
```{r}
#Compute null distribution without sampling
compute_null_distributions <- BRCA2_controls %>%
  rowwise() %>%
  mutate(bal_LAD = log2((CLAD_MMEJ+1)/(CLAD_NHEJ+1)),
                       bal_iLAD = log2((CILAD_MMEJ+1)/(CILAD_NHEJ+1)),
                       bias = bal_LAD - bal_iLAD)

null_summaries <- compute_null_distributions %>% 
  dplyr::group_by(driver,project) %>%
  dplyr::summarise(mean_null = mean(bias),
         sd_null = sd(bias))
ggplot() + geom_density(data= compute_null_distributions, aes(bias)) + geom_density(data = boot_strap_approach, aes(bias), color= "red") + facet_wrap(~ project)



#BRCA2 cases
BRCA2_cases <- mutant_indel_data %>% 
  filter(driver == "BRCA2") %>% 
  reshape2::melt(vars.id = c("TYPE","driver","project")) %>% 
  reshape2::dcast(TYPE+driver+project~chromatin+variable) %>%
  rowwise() %>%
  mutate(bal_LAD = log2((CLAD_MMEJ+1)/(CLAD_NHEJ+1)),
                       bal_iLAD = log2((CILAD_MMEJ+1)/(CILAD_NHEJ+1)),
                       bias = bal_LAD - bal_iLAD) %>%
  left_join(null_summaries)

#compute z-scores
z_score_computation <- BRCA2_cases %>%
  mutate(z_score = (bias - mean_null)/sd_null)


```

