---
title: "xv20230817_mutations_analysis"
author: "x.vergara"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

I will explore the data that I received from Mathijs and proccess it as we did previously for BRCA2mut tumors

```{r}
knitr::opts_chunk$set(root.dir = '/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/rmarkdown/xv20210920_paper_scripts/short_version/knit_html/')

StartTime <-Sys.time()
library(knitr)

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8)

#Change in.dir accordingly
in.dir <- "/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/import/xv20231017_indel_data_mathijs_new/BOOTSTRAP"

#Check if output contain datetag
saveRDS_proof <- function(object, file) {
  if (grepl(Date, file) == T) {
    saveRDS(object, file)
  }
  else {
    print("Incorrect datetag!")
  }
}

#Read rds check through different versions
readRDS_proof <- function(path, pat) {
  full.path <- paste0(in.dir, path)
  document_path <- rstudioapi::getSourceEditorContext()$path %>% str_split(pattern = "/") %>% unlist()
  document_name <- document_path[length(document_path)]
  sys_time <- Sys.time()
  
  if(file.exists(full.path) == T){
    if(paste0(in.dir,path) != getwd()){
      setwd(paste0(in.dir, path))
  }
    file.list <- list.files(pattern = pat)
    correct_file <- str_sort(file.list, decreasing = T)[1]
    print(correct_file)
    write(c(document_name,as.character(sys_time) ,"Input", correct_file), file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
    readRDS(correct_file)
    
  }
  else{
  setwd(path)
  file.list <- list.files(pattern = pat)
  correct_file <- str_sort(file.list, decreasing = T)[1]
  print(correct_file)
  write(c(document_name, as.character(sys_time),"Input", correct_file),file = paste0(in.dir,"log"), ncolumns = 4, sep = "\t", append = T)
  readRDS(correct_file)
  }
}

```

## Libraries:
```{r libraries}
library(readxl)
library(tidyverse)
library(GenomicRanges)
library(rtracklayer)
library(ggbeeswarm)
```

#Import indel data
```{r, warning=F, message=F}
setwd(in.dir)

#Function to import all the data
#List of all driver mutations
driver_mutations_indels <- list.files(path = "/home/x.vergara/XV_P3_ChromDSBScreen/xv_CRISPR_screen_figures/import/xv20231017_indel_data_mathijs_new/BOOTSTRAP", full.names = T)

control_indel_file_names <- map_dfr(driver_mutations_indels, function(y) {
  control_directory <- paste0(y,"/controls_indel")
  #Import all data for a single driver mutation
tumor_file_names <- list.files(path = control_directory, full.names = T)
  mutation_files <- map_dfr(tumor_file_names,function(x) {
    tibble(file_names = list.files(path = x, full.names = T))
  })
  })

#list of files
control_indel_data <- map_dfr(control_indel_file_names$file_names, function(x) {
  read.delim(x) %>%
    mutate(driver = str_extract(x, "(?<=BOOTSTRAP/).*(?=/controls_indel)"),
           project = str_extract(x, "(?<=controls_indel/).*(?=/C)"),
           chromatin = str_extract(x, "CLAD|CILAD"))
})


#Load samples
indel_file_names <- map_dfr(driver_mutations_indels, function(y) {
  #Import all data for a single driver mutation
tumor_file_names <- list.files(path = y, full.names = T)
  mutation_files <- map_dfr(tumor_file_names,function(x) {
    tibble(file_names = list.files(path = x, full.names = T))
  })
  }) %>% filter(grepl(".txt", file_names))

#list of files
mutant_indel_data <- map_dfr(indel_file_names$file_names, function(x) {
  read.delim(`x`) %>%
    mutate(file = str_extract(x, "(?<=BOOTSTRAP/).*LAD")) %>%
    separate(file, into = c("driver","project","chromatin"), sep = "/")
})




```

#Summary numbers controls
```{r}
#Summary table with n
summary_sample_n <- control_indel_data %>%
  dplyr::group_by(driver,project) %>%
  dplyr::summarise(sample_n = n()/2)

#Filtering strategy 
filtered_samples <- control_indel_data %>%
  filter(NHEJ + MMEJ > 2) %>%
  dplyr::group_by(TYPE, driver, project) %>%
  dplyr::summarise(count = n()) %>%
  filter(count == 2) %>%
  ungroup() %>%
    dplyr::group_by(driver,project) %>%
  dplyr::summarise(kept_n = n())

# Filter
ggplot(summary_sample_n) +
  geom_col(aes(project, sample_n, fill = driver), position = "dodge") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))

# Filter out percentage
ggplot(filtered_samples) +
  geom_col(aes(project, kept_n, fill = driver), position = "dodge") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))

#mean control sample are 40
#meadian contol samples are 33

```

#Summary numbers cases
```{r}
#Summary table with n
summary_sample_n_cases <- mutant_indel_data %>%
  dplyr::group_by(driver,project) %>%
  dplyr::summarise(sample_n = n()/2)

#Filtering strategy 
filtered_samples_cases <- mutant_indel_data %>%
  filter(NHEJ + MMEJ > 2) %>%
  dplyr::group_by(TYPE, driver, project) %>%
  dplyr::summarise(count = n()) %>%
  filter(count == 2) %>%
  ungroup() %>%
    dplyr::group_by(driver,project) %>%
  dplyr::summarise(kept_n = n())

# Filter
ggplot(summary_sample_n_cases) +
  geom_col(aes(project, sample_n, fill = driver), position = "dodge") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))

# Filter out percentage
ggplot(filtered_samples_cases) +
  geom_col(aes(project, kept_n, fill = driver), position = "dodge") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))

#mean control sample are 40
#median control samples are 33

```


#What is the percetage of samples discarded
```{r}
discarded_percentage_controls <- summary_sample_n %>% 
  left_join(filtered_samples) %>% 
  mutate(kept_percentage_controls = 100*(kept_n/sample_n))

discarded_percentage_cases <- summary_sample_n_cases %>% 
  left_join(filtered_samples_cases) %>% 
  mutate(kept_percentage_cases = 100*(kept_n/sample_n))

#correlation between them
plot_kept_percentage <-discarded_percentage_controls %>% 
  select(driver,project,kept_percentage_controls) %>% 
  left_join(discarded_percentage_cases%>% 
              select(driver,project,kept_percentage_cases))

#Plot percentage
ggplot(plot_kept_percentage,aes(kept_percentage_controls, kept_percentage_cases)) +
  geom_point() +
  theme_bw() +
  stat_cor() +
  geom_abline() +
  geom_smooth(method = "lm")
  

```

#log2 balance in the control samples (per sample)
```{r, warning=T, message=F}
log2_balance_data_sample <- map_dfr(c(0, 1,2,5,10,20,30,40,50), function(x) {
  control_indel_data %>%
  filter(NHEJ + MMEJ >= x) %>%
  dplyr::group_by(TYPE, driver, project) %>%
  dplyr::summarise(NHEJ = sum(NHEJ),
                   MMEJ = sum(MMEJ)) %>%
  mutate(log2_bal = log2((MMEJ+1)/(NHEJ+1)),
         filter_n = x)
})

#Mean_log2
mean_log2_bal <- log2_balance_data_sample %>%
  dplyr::group_by(driver,project, filter_n) %>%
  dplyr::summarise(mean_log2_bal = mean(log2_bal),
                   sample_n = n(),
                   median_NHEJ = median(NHEJ),
                   median_MMEJ = median(MMEJ))

 ggplot(log2_balance_data_sample,aes(project,log2_bal, color = log2(NHEJ+MMEJ))) + 
   stat_summary(geom = "col") +
   geom_point() +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 90)) + facet_wrap(~ filter_n)
```
#Do I see changes with my mutants (global changes)

#log2 balance in the control samples (per project)
```{r, warning=T, message=F}
log2_balance_data_project <- control_indel_data %>%
  dplyr::group_by(driver, project) %>%
  dplyr::summarise(NHEJ = sum(NHEJ),
                   MMEJ = sum(MMEJ)) %>%
  mutate(mean_log2_bal = log2((MMEJ+1)/(NHEJ+1)))

 ggplot(log2_balance_data_project,aes(project,mean_log2_bal, color = log2(NHEJ+MMEJ))) + 
   stat_summary(geom = "col") +
   geom_point() +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 90))
```

#Plot correlation of those samples
```{r}
summary_mean_controls <- mean_log2_bal %>%
  left_join(log2_balance_data_project %>% select(driver, project, project_sum_mean = mean_log2_bal))

ggplot(summary_mean_controls, aes(mean_log2_bal, project_sum_mean)) + 
  geom_point(aes(size = median_MMEJ)) + 
  stat_cor(method = "spearman") + 
  theme_bw() +
  geom_smooth(method = "lm") +
  xlab("log2_bal of total mutations") +
  ylab("Mean log2_bal of all samples") +
  facet_wrap(~ filter_n)
```
#Conclusion filtering for higher number doesn't help, high amount of mutations actually differs from sum

#Similar approach for chromatin bias
```{r}
chrom_bias_data_sample <- map_dfr(c(0:100), function(x) {
  control_indel_data %>%
  filter(NHEJ + MMEJ >= x) %>%
  dplyr::group_by(TYPE, driver, project,chromatin) %>%
  dplyr::summarise(log2_bal = log2((MMEJ+1)/(NHEJ + 1))) %>%
  reshape2::dcast(TYPE+driver+project ~ chromatin, value.var = "log2_bal") %>%
  mutate(filter_n = x)
}) %>% mutate(bias = CLAD - CILAD)

#Mean_log2
mean_bias <- chrom_bias_data_sample %>%
  dplyr::group_by(driver,project, filter_n) %>%
  dplyr::summarise(mean_bias = mean(bias, na.rm = T),
                   sample_n = n())

 ggplot(chrom_bias_data_sample,aes(project,bias)) + 
   stat_summary(geom = "col") +
#   geom_point() +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 90)) + facet_wrap(~ filter_n)


```

### Mean per tumor project
```{r}
control_data_mean_bias <- control_indel_data %>%
  dplyr::group_by(driver, project,chromatin) %>%
  dplyr::summarise(log2_bal = log2((sum(MMEJ)+1)/(sum(NHEJ) + 1))) %>%
  reshape2::dcast(driver+project ~ chromatin, value.var = "log2_bal")%>% mutate(bias = CLAD - CILAD)

 ggplot(control_data_mean_bias,aes(project,bias)) + 
   stat_summary(geom = "col") +
   geom_point() +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 90))

```

#Correlation between mean and filtered data
```{r}
summary_mean_controls_bias <- mean_bias %>%
  left_join(control_data_mean_bias %>% select(driver, project, project_sum_mean = bias))

#Cortest
summary_cortest <- summary_mean_controls_bias %>%
  dplyr::group_by(filter_n) %>%
  cor_test(mean_bias,project_sum_mean, method = "spearman")

#
ggplot(summary_cortest) + geom_line(aes(filter_n, -log10(p))) +
  theme_bw()

summary_cortest %>% slice_min(p) #10 indels total is the best correlation coefficient

```


```{r}
ggplot(summary_mean_controls_bias %>% filter(filter_n %in% c(0,1,5,10,20,25,50)), aes(mean_bias, project_sum_mean)) + 
  geom_point(aes(size = sample_n)) + 
  stat_cor(method = "spearman") + 
  theme_bw() +
  geom_smooth(method = "lm") +
  xlab("log2_bal of total mutations") +
  ylab("Mean log2_bal of all samples") +
  facet_wrap(~ filter_n)

```
#Conclusion: When it comes to the bias, the mean doesn't resemble to the mean of the tissue

#Measure mean values in mutated samples
```{r}
cases_data_mean_bias <- mutant_indel_data %>%
  dplyr::group_by(driver, project,chromatin) %>%
  dplyr::summarise(log2_bal = log2((sum(MMEJ)+1)/(sum(NHEJ) + 1))) %>%
  reshape2::dcast(driver+project ~ chromatin, value.var = "log2_bal")%>% mutate(bias = CLAD - CILAD)

#Match with controls
control_cases_summary <- cases_data_mean_bias %>%
  select(driver,project,case_bias = bias) %>%
  left_join(control_data_mean_bias %>%
              select(driver, project,control_bias = bias))

```

#Fisher test per tumor class
```{r}
#samples
BRCA2_cases_project <- mutant_indel_data %>% filter(driver == "BRCA2") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver+project~chromatin+variable, fun.aggregate = sum)  %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))
BRCA2_controls_project <- control_indel_data %>% filter(driver == "BRCA2") %>% mutate(driver = "control") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver+project~chromatin+variable, fun.aggregate = sum) %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))

#Plot
BRCA2_total_numbers <- bind_rows(BRCA2_cases_project, BRCA2_controls_project)

ggplot(BRCA2_total_numbers) + geom_quasirandom(aes(fct_relevel(driver, c("control","BRCA2")),bias)) + theme_bw()


wilcox_test <- BRCA2_total_numbers %>%
  wilcox_test(bias ~ driver,paired = T)

```


#BRCA2 total number of mutations (without different projects)
```{r}
#samples
BRCA2_cases_project <- mutant_indel_data %>% filter(driver == "BRCA2") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum)  %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))
BRCA2_controls_project <- control_indel_data %>% filter(driver == "BRCA2") %>% mutate(driver = "control") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum) %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))

#Plot
BRCA2_total_numbers <- bind_rows(BRCA2_cases_project, BRCA2_controls_project)

ggplot(BRCA2_total_numbers) + geom_col(aes(fct_relevel(driver, c("control","BRCA2")),bias)) + theme_bw()

```
#ATM total number of mutations (without different projects)
```{r}
#samples
ATM_cases_project <- mutant_indel_data %>% filter(driver == "ATM") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum)  %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))
ATM_controls_project <- control_indel_data %>% filter(driver == "ATM") %>% mutate(driver = "control") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum) %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))

#Plot
ATM_total_numbers <- bind_rows(ATM_cases_project, ATM_controls_project)

ggplot(ATM_total_numbers) + geom_col(aes(fct_relevel(driver, c("control","ATM")),bias)) + theme_bw() + geom_hline(yintercept = 0, linetype = 2)

```

#EGFR total number of mutations (without different projects)
```{r}
#samples
EGFR_cases_project <- mutant_indel_data %>% filter(driver == "EGFR") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum)  %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))
EGFR_controls_project <- control_indel_data %>% filter(driver == "EGFR") %>% mutate(driver = "control") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(driver~chromatin+variable, fun.aggregate = sum) %>% mutate(bias = log2(CLAD_MMEJ/CLAD_NHEJ) - log2(CILAD_MMEJ/CILAD_NHEJ))

#Plot
EGFR_total_numbers <- bind_rows(EGFR_cases_project, EGFR_controls_project)

ggplot(EGFR_total_numbers) + geom_col(aes(fct_relevel(driver, c("control","ATM")),bias)) + theme_bw() + geom_hline(yintercept = 0, linetype = 2)

```


# Median indel amount per project in controls
```{r}
median_counts <- control_indel_data %>% 
  dplyr::group_by(driver, project, chromatin) %>%
  dplyr::summarise(median_prc = median(NHEJ+MMEJ),
                   n = n()/2)

filter_out_projects <- median_counts %>%
  filter(median_prc < 4) %>% select(driver, project) %>% distinct()

#Mean effect sizes
mean_effects_trends <- control_cases_summary %>% 
  anti_join(filter_out_projects) %>% 
  mutate(delta_effect = case_bias - control_bias)

#Mean effects per gene
mean_driver_effect <-  mean_effects_trends %>%
  dplyr::group_by(driver) %>%
  dplyr::summarise(mean_delta_driver = mean(delta_effect))

```

# Test boot-strapping in controls and build null distributions
```{r}
# Test for BRCA2 first
BRCA2_cases <- mutant_indel_data %>% filter(driver == "BRCA2") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(TYPE+driver+project~chromatin+variable)
BRCA2_controls <- control_indel_data %>% filter(driver == "BRCA2") %>% reshape2::melt(vars.id = c("TYPE","driver","project")) %>% reshape2::dcast(TYPE+driver+project~chromatin+variable)

#Number of cases per tumor type 
n_cases_BRCA2 <- BRCA2_cases %>%
  select(TYPE,driver,project) %>%
  distinct() %>%
  dplyr::group_by(driver, project) %>% 
  dplyr::summarise(n = n())

#Number of controls per each tumor type (as QC)
n_controls_BRCA2 <- BRCA2_controls %>%
  select(TYPE,driver,project) %>%
  distinct() %>%
  dplyr::group_by(driver, project) %>% 
  dplyr::summarise(n = n())

#Bootstrap approach (before it repeats)
boot_strap_approach <- map2_dfr(n_cases_BRCA2$driver,n_cases_BRCA2$project, function(x,y){
  project_file <- BRCA2_controls %>% filter(driver == x & project == y)
  n_drawing <- n_cases_BRCA2 %>% filter(driver == x & project == y) %>% pull(n) %>% unique()
  map_dfr(c(1:2000), function(z){
    set.seed(z)
    project_file %>% 
      sample_n(n_drawing) %>%
      mutate(iteration = z) %>%
      dplyr::group_by(iteration) %>%
      dplyr::summarise(bal_LAD = log2((sum(CLAD_MMEJ))/(sum(CLAD_NHEJ))),
                       bal_iLAD = log2((sum(CILAD_MMEJ))/(sum(CILAD_NHEJ))),
                       bias = bal_LAD - bal_iLAD) %>%
      mutate(driver = x, project = y, sampling_n = n_drawing)
  })
})

#Calculate mean and sd of each approach
tmp <- boot_strap_approach %>%
  filter(is.finite(bias)) %>% dplyr::group_by(driver,project, sampling_n) %>%
  dplyr::summarise(mean_null = mean(bias, na.rm =T),
                  sd_null = sd(bias, na.rm = T)) %>% left_join(mean_effects_trends) %>%
  mutate(z_score = (case_bias - mean_null)/sd_null)

#Compute metric before
BRCA2_controls_metric <- BRCA2_controls %>% 
  rowwise() %>% 
  mutate(bal_LAD = log2((CLAD_MMEJ+1)/(CLAD_NHEJ+1)),
         bal_iLAD = log2((CILAD_MMEJ+1)/(CILAD_NHEJ+1)),
         bias = bal_LAD - bal_iLAD)

boot_strap_approach_mean <- map2_dfr(n_cases_BRCA2$driver,n_cases_BRCA2$project, function(x,y){
  project_file <- BRCA2_controls_metric %>% filter(driver == x & project == y)
#  n_drawing <- n_cases_BRCA2 %>% filter(driver == x & project == y) %>% pull(n) %>% unique()
#  map_dfr(c(1:100), function(z){
#    set.seed(z)
#    mean_bias <- project_file %>% 
#      sample_n(n_drawing) %>%
#      pull(bias) %>%
#      mean()
#    
#    tibble(driver = x, project = y, avg_bias = mean_bias,iteraction = z, n_sampling = n_drawing)
#  })
})



```

#Compute mean and sd per project
```{r}
#Compute null distribution without sampling
compute_null_distributions <- BRCA2_controls %>%
  rowwise() %>%
  mutate(bal_LAD = log2((CLAD_MMEJ+1)/(CLAD_NHEJ+1)),
                       bal_iLAD = log2((CILAD_MMEJ+1)/(CILAD_NHEJ+1)),
                       bias = bal_LAD - bal_iLAD)

null_summaries <- compute_null_distributions %>% 
  dplyr::group_by(driver,project) %>%
  dplyr::summarise(mean_null = mean(bias),
         sd_null = sd(bias))
ggplot() + geom_density(data= compute_null_distributions, aes(bias)) + geom_density(data = boot_strap_approach, aes(bias), color= "red") + facet_wrap(~ project)



#BRCA2 cases
BRCA2_cases <- mutant_indel_data %>% 
  filter(driver == "BRCA2") %>% 
  reshape2::melt(vars.id = c("TYPE","driver","project")) %>% 
  reshape2::dcast(TYPE+driver+project~chromatin+variable) %>%
  rowwise() %>%
  mutate(bal_LAD = log2((CLAD_MMEJ+1)/(CLAD_NHEJ+1)),
                       bal_iLAD = log2((CILAD_MMEJ+1)/(CILAD_NHEJ+1)),
                       bias = bal_LAD - bal_iLAD) %>%
  left_join(null_summaries)

#compute z-scores
z_score_computation <- BRCA2_cases %>%
  mutate(z_score = (bias - mean_null)/sd_null)


```

